// tdesign-react/cjs/table/base/TableAsyncLoadingBody.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var isFunction = require('../../_chunks/dep-1bafebb1.js');
var loading_index = require('../../loading/index.js');
var _util_useConfig = require('../../_util/useConfig.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-735b841a.js');
require('../../loading/loading.js');
require('../../_chunks/dep-449daca3.js');
require('../../_chunks/dep-fa36e79f.js');
require('../../_chunks/dep-045b87c1.js');
require('classnames');
require('../../_util/dom.js');
require('raf');
require('../../_chunks/dep-0705da20.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_util/easing.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/gradient.js');
require('../../config-provider/ConfigContext.js');
require('../../locale/zh_CN.js');
require('../../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function TableEmptyBody(props) {
  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var asyncLoading = props.asyncLoading,
      onAsyncLoadingClick = props.onAsyncLoadingClick;
  if (!asyncLoading) return null;

  if (asyncLoading === "loading") {
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-table__async-loading ").concat(classPrefix, "-is-loading"),
      onClick: function onClick() {
        onAsyncLoadingClick({
          status: "loading"
        });
      }
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
      loading: true,
      size: "small"
    }), /* @__PURE__ */React__default["default"].createElement("span", null, "\u6B63\u5728\u52A0\u8F7D\u4E2D\uFF0C\u8BF7\u7A0D\u540E"));
  }

  if (asyncLoading === "load-more") {
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-table__async-loading ").concat(classPrefix, "-is-load-more"),
      onClick: function onClick() {
        onAsyncLoadingClick({
          status: "load-more"
        });
      }
    }, "\u70B9\u51FB\u52A0\u8F7D\u66F4\u591A");
  }

  var result = null;
  if (asyncLoading) result = asyncLoading;
  if (isFunction.isFunction_1(asyncLoading)) result = asyncLoading();
  return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, result);
}

exports["default"] = TableEmptyBody;
