// tdesign-react/lib/time-picker/consts.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-93971e86.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import '../_chunks/dep-beac373a.js';
import '../_chunks/dep-9332bbea.js';
import 'react';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';

var useTimePickerTextConfig = function useTimePickerTextConfig() {
  var _useLocaleReceiver = useLocaleReceiver("timePicker"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  return {
    nowtime: t(local.nowtime),
    confirm: t(local.confirm),
    am: t(local.anteMeridiem),
    pm: t(local.postMeridiem),
    placeholder: t(local.placeholder)
  };
};
var AM = "am";
var PM = "pm";
var MERIDIEM_LIST = [AM, PM];
var DEFAULT_STEPS = [1, 1, 1];
var DEFAULT_FORMAT = "HH:mm:ss";

export { AM, DEFAULT_FORMAT, DEFAULT_STEPS, MERIDIEM_LIST, PM, useTimePickerTextConfig };
