// tdesign-react/lib/upload/themes/flow-list/img-list.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-9332bbea.js';
import classNames from 'classnames';
import { ErrorCircleFilledIcon, BrowseIcon, DeleteIcon } from 'tdesign-icons-react';
import React from 'react';
import useConfig from '../../../_util/useConfig.js';
import { abridgeName } from '../../util.js';
import { Loading } from '../../../loading/index.js';
import '../../../config-provider/ConfigContext.js';
import '../../../locale/zh_CN.js';
import '../../../_chunks/dep-124f91c5.js';
import '../../../_chunks/dep-beac373a.js';
import '../../../loading/loading.js';
import '../../../_chunks/dep-93971e86.js';
import '../../../_util/dom.js';
import 'raf';
import '../../../_chunks/dep-f3f6eedc.js';
import '../../../_chunks/dep-c6576f8f.js';
import '../../../_chunks/dep-156c830d.js';
import '../../../_chunks/dep-04305406.js';
import '../../../_chunks/dep-6ca758b4.js';
import '../../../_util/easing.js';
import '../../../common/Portal.js';
import 'react-dom';
import '../../../loading/gradient.js';
import '../../../loading/plugin.js';

var ImgList = function ImgList(props) {
  var showInitial = props.showInitial,
      listFiles = props.listFiles,
      renderDragger = props.renderDragger,
      onImgPreview = props.onImgPreview,
      remove = props.remove;

  var _useConfig = useConfig(),
      prefix = _useConfig.classPrefix;

  var UPLOAD_NAME = "".concat(prefix, "-upload");
  var UPLOAD_NAME_CARD = "".concat(UPLOAD_NAME, "__card");
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-card-area")
  }, showInitial && renderDragger(), !!listFiles.length && /* @__PURE__ */React.createElement("ul", {
    className: "".concat(UPLOAD_NAME_CARD, " clearfix")
  }, listFiles.map(function (file, index) {
    var _classNames;

    return /* @__PURE__ */React.createElement("li", {
      className: "".concat(UPLOAD_NAME_CARD, "-item"),
      key: index
    }, /* @__PURE__ */React.createElement("div", {
      className: classNames((_classNames = {}, _defineProperty(_classNames, "".concat(UPLOAD_NAME_CARD, "-content"), true), _defineProperty(_classNames, "".concat(UPLOAD_NAME_CARD, "-content-border"), file.status !== "waiting"), _classNames))
    }, file.status === "fail" && /* @__PURE__ */React.createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-status-wrap")
    }, /* @__PURE__ */React.createElement(ErrorCircleFilledIcon, null), /* @__PURE__ */React.createElement("p", null, "\u4E0A\u4F20\u5931\u8D25")), file.status === "progress" && /* @__PURE__ */React.createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-status-wrap")
    }, /* @__PURE__ */React.createElement(Loading, null), /* @__PURE__ */React.createElement("p", null, "\u4E0A\u4F20\u4E2D ", Math.min(file.percent, 99))), (["waiting", "success"].includes(file.status) || !file.status && file.url) && /* @__PURE__ */React.createElement("img", {
      className: "".concat(UPLOAD_NAME_CARD, "-image"),
      src: file.url || "//tdesign.gtimg.com/tdesign-default-img.png",
      alt: ""
    }), /* @__PURE__ */React.createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-mask")
    }, file.url && /* @__PURE__ */React.createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "__mask__item")
    }, /* @__PURE__ */React.createElement(BrowseIcon, {
      onClick: function onClick(e) {
        return onImgPreview(e, file);
      }
    }), /* @__PURE__ */React.createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "__mask__item-divider")
    })), /* @__PURE__ */React.createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "-mask__item"),
      onClick: function onClick(e) {
        return remove({
          e: e,
          index: index,
          file: file
        });
      }
    }, /* @__PURE__ */React.createElement(DeleteIcon, null)))), /* @__PURE__ */React.createElement("p", {
      className: "".concat(UPLOAD_NAME_CARD, "-name")
    }, abridgeName(file.name)));
  })));
};

export { ImgList as default };
