// tdesign-react/lib/upload/themes/flow-list/index.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-9332bbea.js';
import { _ as _slicedToArray } from '../../../_chunks/dep-93971e86.js';
import React, { useState } from 'react';
import classNames from 'classnames';
import useConfig from '../../../_util/useConfig.js';
import { Button } from '../../../button/index.js';
import ImgList from './img-list.js';
import FileList from './file-list.js';
import BooleanRender from '../../boolean-render.js';
import '../../../_chunks/dep-beac373a.js';
import '../../../config-provider/ConfigContext.js';
import '../../../locale/zh_CN.js';
import '../../../button/Button.js';
import '../../../_chunks/dep-27d690ce.js';
import '../../../_util/noop.js';
import '../../../_util/useRipple.js';
import '../../../_util/setStyle.js';
import '../../../loading/index.js';
import '../../../loading/loading.js';
import '../../../_util/dom.js';
import 'raf';
import '../../../_chunks/dep-f3f6eedc.js';
import '../../../_chunks/dep-c6576f8f.js';
import '../../../_chunks/dep-156c830d.js';
import '../../../_chunks/dep-04305406.js';
import '../../../_chunks/dep-6ca758b4.js';
import '../../../_util/easing.js';
import '../../../common/Portal.js';
import 'react-dom';
import '../../../loading/gradient.js';
import '../../../loading/plugin.js';
import 'tdesign-icons-react';
import '../../util.js';
import '../../../_chunks/dep-124f91c5.js';

var Index = function Index(props) {
  var _classNames;

  var listFiles = props.files,
      _props$toUploadFiles = props.toUploadFiles,
      toUploadFiles = _props$toUploadFiles === void 0 ? [] : _props$toUploadFiles,
      _props$showUploadProg = props.showUploadProgress,
      showUploadProgress = _props$showUploadProg === void 0 ? false : _props$showUploadProg,
      placeholder = props.placeholder,
      display = props.display,
      onImgPreview = props.onImgPreview,
      remove = props.remove,
      upload = props.upload,
      cancel = props.cancel,
      onChange = props.onChange,
      onDragenter = props.onDragenter,
      onDragleave = props.onDragleave,
      _props$children = props.children,
      children = _props$children === void 0 ? null : _props$children;
  var target = React.useRef();

  var _useConfig = useConfig(),
      prefix = _useConfig.classPrefix;

  var UPLOAD_NAME = "".concat(prefix, "-upload");

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      dragActive = _useState2[0],
      setDragActive = _useState2[1];

  var showInitial = !listFiles.length;
  var failedList = toUploadFiles.filter(function (file) {
    return file.status === "fail";
  });
  var isUploading = toUploadFiles.filter(function (file) {
    return file.status === "progress";
  }).length > 0;
  var allowUpload = toUploadFiles.length > 0 && !isUploading;
  var uploadText = failedList.length ? "\u91CD\u65B0\u4E0A\u4F20" : "\u5F00\u59CB\u4E0A\u4F20";
  if (isUploading) uploadText = "\u4E0A\u4F20\u4E2D...";

  var handleDrop = function handleDrop(event) {
    event.preventDefault();
    onChange === null || onChange === void 0 ? void 0 : onChange(event.dataTransfer.files);
    onDragleave === null || onDragleave === void 0 ? void 0 : onDragleave(event);
    setDragActive(false);
  };

  var handleDragenter = function handleDragenter(event) {
    event.preventDefault();
    onDragenter === null || onDragenter === void 0 ? void 0 : onDragenter(event);
    setDragActive(true);
  };

  var handleDragleave = function handleDragleave(event) {
    if (event.target !== target.current) return;
    event.preventDefault();
    onDragleave === null || onDragleave === void 0 ? void 0 : onDragleave(event);
    setDragActive(false);
  };

  var handleDragover = function handleDragover(event) {
    event.preventDefault();
  };

  var renderDragger = function renderDragger() {
    return /* @__PURE__ */React.createElement("div", {
      ref: target,
      className: "".concat(UPLOAD_NAME, "__flow-empty"),
      onDrop: handleDrop,
      onDragEnter: handleDragenter,
      onDragOver: handleDragover,
      onDragLeave: handleDragleave
    }, dragActive ? "\u91CA\u653E\u9F20\u6807" : "\u70B9\u51FB\u4E0A\u65B9\u201C\u9009\u62E9\u6587\u4EF6\u201D\u6216\u5C06\u6587\u4EF6\u62D6\u62FD\u5230\u6B64\u533A\u57DF");
  };

  var wrapperClassNames = classNames((_classNames = {}, _defineProperty(_classNames, "".concat(UPLOAD_NAME, "__flow"), true), _defineProperty(_classNames, "".concat(UPLOAD_NAME, "__flow-").concat(display), true), _classNames));
  return /* @__PURE__ */React.createElement("div", {
    className: wrapperClassNames
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-op")
  }, children, /* @__PURE__ */React.createElement("small", {
    className: "".concat(UPLOAD_NAME, "__small ").concat(UPLOAD_NAME, "__flow-placeholder")
  }, placeholder)), /* @__PURE__ */React.createElement(BooleanRender, {
    boolExpression: display === "file-flow"
  }, /* @__PURE__ */React.createElement(FileList, {
    listFiles: listFiles,
    showInitial: showInitial,
    renderDragger: renderDragger,
    showUploadProgress: showUploadProgress,
    remove: remove
  })), /* @__PURE__ */React.createElement(BooleanRender, {
    boolExpression: display === "image-flow"
  }, /* @__PURE__ */React.createElement(ImgList, {
    listFiles: listFiles,
    showInitial: showInitial,
    renderDragger: renderDragger,
    onImgPreview: onImgPreview,
    remove: remove
  })), /* @__PURE__ */React.createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-bottom")
  }, /* @__PURE__ */React.createElement(Button, {
    theme: "default",
    onClick: cancel
  }, "\u53D6\u6D88"), /* @__PURE__ */React.createElement(Button, {
    disabled: !allowUpload,
    theme: "primary",
    onClick: function onClick(e) {
      return upload(toUploadFiles, e);
    }
  }, uploadText)));
};

export { Index as default };
