// tdesign-react/cjs/input-number/StepHandler.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-449daca3.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var _util_useCommonClassName = require('../_util/useCommonClassName.js');
var button_index = require('../button/index.js');
require('../_util/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../locale/zh_CN.js');
require('../button/Button.js');
require('../_chunks/dep-e03ca79d.js');
require('../_util/noop.js');
require('../_util/useRipple.js');
require('../_util/setStyle.js');
require('../loading/index.js');
require('../loading/loading.js');
require('../_chunks/dep-fa36e79f.js');
require('../_chunks/dep-045b87c1.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-0705da20.js');
require('../_chunks/dep-0e3f1753.js');
require('../_chunks/dep-a7666a95.js');
require('../_chunks/dep-9125a75e.js');
require('../_chunks/dep-48870ae5.js');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function StepHandler(props) {
  var prefixClassName = props.prefixClassName,
      theme = props.theme,
      onStep = props.onStep,
      disabledDecrease = props.disabledDecrease,
      disabledIncrease = props.disabledIncrease;
  var commonClassNames = _util_useCommonClassName["default"]();
  var decreaseIcon = theme === "column" ? /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ChevronDownIcon, null) : /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.RemoveIcon, null);
  var increaseIcon = theme === "column" ? /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ChevronUpIcon, null) : /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.AddIcon, null);

  var onStepDecrease = function onStepDecrease(e) {
    return disabledDecrease || onStep({
      type: "reduce",
      e: e
    });
  };

  var onStepIncrease = function onStepIncrease(e) {
    return disabledIncrease || onStep({
      type: "add",
      e: e
    });
  };

  return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    variant: "outline",
    className: classNames__default["default"]("".concat(prefixClassName, "__decrease"), defineProperty._defineProperty({}, commonClassNames.STATUS.disabled, disabledDecrease)),
    onClick: onStepDecrease,
    icon: decreaseIcon,
    shape: "square"
  }), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    variant: "outline",
    className: classNames__default["default"]("".concat(prefixClassName, "__increase"), defineProperty._defineProperty({}, commonClassNames.STATUS.disabled, disabledIncrease)),
    onClick: onStepIncrease,
    icon: increaseIcon,
    shape: "square"
  }));
}

exports["default"] = StepHandler;
