// tdesign-react/cjs/upload/themes/flow-list/file-list.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var upload_util = require('../../util.js');
var _util_useConfig = require('../../../_util/useConfig.js');
var loading_index = require('../../../loading/index.js');
require('../../../_chunks/dep-8f3deb2a.js');
require('../../../_chunks/dep-045b87c1.js');
require('../../../config-provider/ConfigContext.js');
require('../../../locale/zh_CN.js');
require('../../../loading/loading.js');
require('../../../_chunks/dep-449daca3.js');
require('../../../_chunks/dep-fa36e79f.js');
require('classnames');
require('../../../_util/dom.js');
require('raf');
require('../../../_chunks/dep-0705da20.js');
require('../../../_chunks/dep-0e3f1753.js');
require('../../../_chunks/dep-a7666a95.js');
require('../../../_chunks/dep-9125a75e.js');
require('../../../_chunks/dep-48870ae5.js');
require('../../../_util/easing.js');
require('../../../common/Portal.js');
require('react-dom');
require('../../../loading/gradient.js');
require('../../../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var FileList = function FileList(props) {
  var listFiles = props.listFiles,
      showInitial = props.showInitial,
      renderDragger = props.renderDragger,
      showUploadProgress = props.showUploadProgress,
      remove = props.remove;

  var _useConfig = _util_useConfig["default"](),
      prefix = _useConfig.classPrefix;

  var UPLOAD_NAME = "".concat(prefix, "-upload");

  var renderStatus = function renderStatus(file) {
    var STATUS_MAP = {
      success: {
        icon: /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CheckCircleFilledIcon, null),
        text: "\u4E0A\u4F20\u6210\u529F"
      },
      fail: {
        icon: /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ErrorCircleFilledIcon, null),
        text: "\u4E0A\u4F20\u5931\u8D25"
      },
      progress: {
        icon: /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, null),
        text: "\u4E0A\u4F20\u4E2D ".concat(Math.min(file.percent, 99), "%")
      },
      waiting: {
        icon: /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.TimeFilledIcon, null),
        text: "\u5F85\u4E0A\u4F20"
      }
    };
    if (file.status === "progress" && !showUploadProgress) return null;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME, "__flow-status")
    }, STATUS_MAP[file.status].icon, /* @__PURE__ */React__default["default"].createElement("span", null, STATUS_MAP[file.status].text));
  };

  return /* @__PURE__ */React__default["default"].createElement("table", {
    className: "".concat(UPLOAD_NAME, "__flow-table")
  }, /* @__PURE__ */React__default["default"].createElement("thead", null, /* @__PURE__ */React__default["default"].createElement("tr", null, /* @__PURE__ */React__default["default"].createElement("th", null, "\u6587\u4EF6\u540D"), /* @__PURE__ */React__default["default"].createElement("th", null, "\u5927\u5C0F"), /* @__PURE__ */React__default["default"].createElement("th", null, "\u72B6\u6001"), /* @__PURE__ */React__default["default"].createElement("th", null, "\u64CD\u4F5C"))), /* @__PURE__ */React__default["default"].createElement("tbody", null, showInitial && /* @__PURE__ */React__default["default"].createElement("tr", null, /* @__PURE__ */React__default["default"].createElement("td", {
    colSpan: 4
  }, renderDragger())), listFiles.map(function (file, index) {
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      key: index
    }, /* @__PURE__ */React__default["default"].createElement("td", null, upload_util.abridgeName(file.name, 7, 10)), /* @__PURE__ */React__default["default"].createElement("td", null, upload_util.returnFileSize(file.size)), /* @__PURE__ */React__default["default"].createElement("td", null, renderStatus(file)), /* @__PURE__ */React__default["default"].createElement("td", null, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME, "__flow-button"),
      onClick: function onClick(e) {
        return remove({
          e: e,
          index: index,
          file: file
        });
      }
    }, "\u5220\u9664")));
  })));
};

exports["default"] = FileList;
