// tdesign-react/cjs/upload/themes/flow-list/img-list.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-449daca3.js');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var React = require('react');
var _util_useConfig = require('../../../_util/useConfig.js');
var upload_util = require('../../util.js');
var loading_index = require('../../../loading/index.js');
require('../../../config-provider/ConfigContext.js');
require('../../../locale/zh_CN.js');
require('../../../_chunks/dep-8f3deb2a.js');
require('../../../_chunks/dep-045b87c1.js');
require('../../../loading/loading.js');
require('../../../_chunks/dep-fa36e79f.js');
require('../../../_util/dom.js');
require('raf');
require('../../../_chunks/dep-0705da20.js');
require('../../../_chunks/dep-0e3f1753.js');
require('../../../_chunks/dep-a7666a95.js');
require('../../../_chunks/dep-9125a75e.js');
require('../../../_chunks/dep-48870ae5.js');
require('../../../_util/easing.js');
require('../../../common/Portal.js');
require('react-dom');
require('../../../loading/gradient.js');
require('../../../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var ImgList = function ImgList(props) {
  var showInitial = props.showInitial,
      listFiles = props.listFiles,
      renderDragger = props.renderDragger,
      onImgPreview = props.onImgPreview,
      remove = props.remove;

  var _useConfig = _util_useConfig["default"](),
      prefix = _useConfig.classPrefix;

  var UPLOAD_NAME = "".concat(prefix, "-upload");
  var UPLOAD_NAME_CARD = "".concat(UPLOAD_NAME, "__card");
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-card-area")
  }, showInitial && renderDragger(), !!listFiles.length && /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(UPLOAD_NAME_CARD, " clearfix")
  }, listFiles.map(function (file, index) {
    var _classNames;

    return /* @__PURE__ */React__default["default"].createElement("li", {
      className: "".concat(UPLOAD_NAME_CARD, "-item"),
      key: index
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]((_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(UPLOAD_NAME_CARD, "-content"), true), defineProperty._defineProperty(_classNames, "".concat(UPLOAD_NAME_CARD, "-content-border"), file.status !== "waiting"), _classNames))
    }, file.status === "fail" && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-status-wrap")
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ErrorCircleFilledIcon, null), /* @__PURE__ */React__default["default"].createElement("p", null, "\u4E0A\u4F20\u5931\u8D25")), file.status === "progress" && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-status-wrap")
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, null), /* @__PURE__ */React__default["default"].createElement("p", null, "\u4E0A\u4F20\u4E2D ", Math.min(file.percent, 99))), (["waiting", "success"].includes(file.status) || !file.status && file.url) && /* @__PURE__ */React__default["default"].createElement("img", {
      className: "".concat(UPLOAD_NAME_CARD, "-image"),
      src: file.url || "//tdesign.gtimg.com/tdesign-default-img.png",
      alt: ""
    }), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-mask")
    }, file.url && /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "__mask__item")
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.BrowseIcon, {
      onClick: function onClick(e) {
        return onImgPreview(e, file);
      }
    }), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "__mask__item-divider")
    })), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "-mask__item"),
      onClick: function onClick(e) {
        return remove({
          e: e,
          index: index,
          file: file
        });
      }
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.DeleteIcon, null)))), /* @__PURE__ */React__default["default"].createElement("p", {
      className: "".concat(UPLOAD_NAME_CARD, "-name")
    }, upload_util.abridgeName(file.name)));
  })));
};

exports["default"] = ImgList;
