// tdesign-react/cjs/notification/Notification.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-fa36e79f.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var _util_noop = require('../_util/noop.js');
var _util_useConfig = require('../_util/useConfig.js');
require('../_chunks/dep-045b87c1.js');
require('../config-provider/ConfigContext.js');
require('../locale/zh_CN.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var blockName = "notification";
var NotificationComponent = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$title = props.title,
      title = _props$title === void 0 ? null : _props$title,
      _props$content = props.content,
      content = _props$content === void 0 ? null : _props$content,
      _props$theme = props.theme,
      theme = _props$theme === void 0 ? null : _props$theme,
      _props$icon = props.icon,
      icon = _props$icon === void 0 ? null : _props$icon,
      closeBtn = props.closeBtn,
      _props$footer = props.footer,
      footer = _props$footer === void 0 ? null : _props$footer,
      _props$duration = props.duration,
      duration = _props$duration === void 0 ? 3e3 : _props$duration,
      _props$onCloseBtnClic = props.onCloseBtnClick,
      onCloseBtnClick = _props$onCloseBtnClic === void 0 ? _util_noop["default"] : _props$onCloseBtnClic,
      _props$onDurationEnd = props.onDurationEnd,
      onDurationEnd = _props$onDurationEnd === void 0 ? _util_noop["default"] : _props$onDurationEnd,
      style = props.style;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var prefixCls = React__default["default"].useCallback(function () {
    var className = "";

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    args.forEach(function (item, index) {
      if (item && index > 0) className = className.concat(" ");

      if (item instanceof Array) {
        var _item = slicedToArray._slicedToArray(item, 3),
            block = _item[0],
            element = _item[1],
            modifier = _item[2];

        className = className.concat(classPrefix, "-", block);
        if (element) className = className.concat("__", element);
        if (modifier) className = className.concat("--", modifier);
      } else if (typeof item === "string") {
        className = className.concat(classPrefix, "-", item);
      }
    });
    return className;
  }, [classPrefix]);
  React__default["default"].useImperativeHandle(ref, function () {
    return {
      close: close
    };
  }, []);
  React__default["default"].useEffect(function () {
    var timer;

    if (duration > 0) {
      timer = setTimeout(function () {
        clearTimeout(timer);
        onDurationEnd();
      }, duration);
    }

    return function () {
      if (timer) clearTimeout(timer);
    };
  }, []);

  var renderIcon = function renderIcon() {
    var IconWrapper = function IconWrapper(_ref) {
      var children = _ref.children;
      return /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-notification__icon")
      }, children);
    };

    if ( /*#__PURE__*/React__default["default"].isValidElement(icon)) {
      return /* @__PURE__ */React__default["default"].createElement(IconWrapper, null, icon);
    }

    if (theme && theme === "success") {
      return /* @__PURE__ */React__default["default"].createElement(IconWrapper, null, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CheckCircleFilledIcon, {
        className: prefixCls("is-success")
      }));
    }

    if (theme && ["info", "warning", "error"].indexOf(theme) >= 0) {
      return /* @__PURE__ */React__default["default"].createElement(IconWrapper, null, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.InfoCircleFilledIcon, {
        className: prefixCls("is-".concat(theme))
      }));
    }

    return null;
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    className: prefixCls(blockName),
    style: style
  }, renderIcon(), /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([blockName, "main"])
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([blockName, "title__wrap"])
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: prefixCls([blockName, "title"])
  }, title), function () {
    if (typeof closeBtn === "boolean" && closeBtn) {
      return /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CloseIcon, {
        className: prefixCls("icon-close"),
        onClick: function onClick(e) {
          onCloseBtnClick({
            e: e
          });
        }
      });
    }

    if ( /*#__PURE__*/React__default["default"].isValidElement(closeBtn)) {
      return /* @__PURE__ */React__default["default"].createElement("div", {
        onClick: function onClick(e) {
          onCloseBtnClick({
            e: e
          });
        }
      }, closeBtn);
    }

    return null;
  }()), function () {
    if (typeof content === "string") {
      return /* @__PURE__ */React__default["default"].createElement("div", {
        className: prefixCls([blockName, "content"])
      }, content);
    }

    if ( /*#__PURE__*/React__default["default"].isValidElement(content)) return content;
    return null;
  }(), /*#__PURE__*/React__default["default"].isValidElement(footer) && /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([blockName, "detail"])
  }, footer), typeof footer === "function" && /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([blockName, "detail"])
  }, footer())));
});

exports.NotificationComponent = NotificationComponent;
exports["default"] = NotificationComponent;
