// tdesign-react/cjs/notification/NotificationList.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-449daca3.js');
var toConsumableArray = require('../_chunks/dep-8f3deb2a.js');
var slicedToArray = require('../_chunks/dep-fa36e79f.js');
var React = require('react');
var ReactDOM = require('react-dom');
var _util_useConfig = require('../_util/useConfig.js');
var _util_noop = require('../_util/noop.js');
var notification_Notification = require('./Notification.js');
require('../_chunks/dep-045b87c1.js');
require('../config-provider/ConfigContext.js');
require('../locale/zh_CN.js');
require('tdesign-icons-react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var ReactDOM__default = /*#__PURE__*/_interopDefaultLegacy(ReactDOM);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var seed = 0;
var listMap = /* @__PURE__ */new Map();
var NotificationList = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var attach = props.attach,
      placement = props.placement,
      zIndex = props.zIndex;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _React$useReducer = React__default["default"].useReducer(function (state, action) {
    switch (action.type) {
      case "push":
        return [].concat(toConsumableArray._toConsumableArray(state), [action.value]);

      case "remove":
        return state.filter(function (item) {
          return item.key !== action.key;
        }).map(function (item) {
          return item;
        });

      case "removeAll":
        return [];

      default:
        return state;
    }
  }, []),
      _React$useReducer2 = slicedToArray._slicedToArray(_React$useReducer, 2),
      list = _React$useReducer2[0],
      dispatchList = _React$useReducer2[1];

  var notificationMap = React__default["default"].useMemo(function () {
    return /* @__PURE__ */new Map();
  }, []);
  var remove = React__default["default"].useCallback(function (key) {
    dispatchList({
      type: "remove",
      key: key
    });
    notificationMap["delete"](key);
  }, [notificationMap]);

  var calOffset = function calOffset(offset) {
    if (!offset) return "16px";
    return isNaN(Number(offset)) ? offset : "".concat(offset, "px");
  };

  var push = React__default["default"].useCallback(function (theme, options) {
    return new Promise(function (resolve) {
      var key = String(seed += 1);

      var style = function () {
        if (Array.isArray(options.offset)) {
          var _ref = toConsumableArray._toConsumableArray(options.offset),
              horizontal = _ref[0],
              vertical = _ref[1];

          var horizontalOffset = calOffset(horizontal);
          var verticalOffset = calOffset(vertical);
          return {
            marginTop: verticalOffset,
            marginBottom: verticalOffset,
            marginLeft: horizontalOffset,
            marginRight: horizontalOffset
          };
        }

        return {
          margin: "16px"
        };
      }();

      notificationMap.set(key, /*#__PURE__*/React__default["default"].createRef());
      dispatchList({
        type: "push",
        value: _objectSpread(_objectSpread({}, options), {}, {
          key: key,
          theme: theme,
          style: style
        })
      });

      notificationMap.get(key).current.close = function () {
        remove(key);
      };

      resolve(notificationMap.get(key).current);
    });
  }, [notificationMap, remove]);
  var removeAll = React__default["default"].useCallback(function () {
    dispatchList({
      type: "removeAll"
    });
    notificationMap.clear();
  }, [notificationMap]);
  React__default["default"].useImperativeHandle(ref, function () {
    return {
      push: push,
      remove: remove,
      removeAll: removeAll
    };
  });
  React__default["default"].useEffect(function () {
    if (list.length === 0 && notificationMap.size === 0) {
      listMap["delete"](placement);
      ReactDOM__default["default"].unmountComponentAtNode(attach);
      attach.remove();
    }
  }, [list, attach, placement, notificationMap]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-notification__show--").concat(placement),
    style: {
      zIndex: zIndex
    }
  }, list.map(function (props2) {
    var _props2$onDurationEnd = props2.onDurationEnd,
        _onDurationEnd = _props2$onDurationEnd === void 0 ? _util_noop["default"] : _props2$onDurationEnd,
        _props2$onCloseBtnCli = props2.onCloseBtnClick,
        _onCloseBtnClick = _props2$onCloseBtnCli === void 0 ? _util_noop["default"] : _props2$onCloseBtnCli;

    return /* @__PURE__ */React__default["default"].createElement(notification_Notification.NotificationComponent, _objectSpread(_objectSpread({
      theme: "warning",
      ref: notificationMap.get(props2.key),
      key: props2.key
    }, props2), {}, {
      onDurationEnd: function onDurationEnd() {
        remove(props2.key);

        _onDurationEnd();
      },
      onCloseBtnClick: function onCloseBtnClick(e) {
        remove(props2.key);

        _onCloseBtnClick(e);
      }
    }));
  }));
});
var fetchListInstance = function fetchListInstance(placement, attach, zIndex) {
  return new Promise(function (resolve) {
    if (listMap.has(placement)) {
      resolve(listMap.get(placement));
    } else {
      var hasExec = false;
      ReactDOM__default["default"].render( /* @__PURE__ */React__default["default"].createElement(NotificationList, {
        attach: attach,
        placement: placement,
        zIndex: Number(zIndex),
        ref: function ref(instance) {
          if (!hasExec) {
            hasExec = true;
            listMap.set(placement, instance);
            resolve(instance);
          }
        }
      }), attach);
    }
  });
};

exports.fetchListInstance = fetchListInstance;
exports.listMap = listMap;
