// tdesign-react/cjs/upload/themes/dragger-progress.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var button_index = require('../../button/index.js');
var loading_index = require('../../loading/index.js');
var _util_useConfig = require('../../_util/useConfig.js');
var upload_util = require('../util.js');
require('../../button/Button.js');
require('../../_chunks/dep-449daca3.js');
require('../../_chunks/dep-e03ca79d.js');
require('classnames');
require('../../_util/noop.js');
require('../../_util/useRipple.js');
require('../../_util/setStyle.js');
require('../../config-provider/ConfigContext.js');
require('../../locale/zh_CN.js');
require('../../loading/loading.js');
require('../../_chunks/dep-fa36e79f.js');
require('../../_chunks/dep-045b87c1.js');
require('../../_util/dom.js');
require('raf');
require('../../_chunks/dep-0705da20.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_util/easing.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/gradient.js');
require('../../loading/plugin.js');
require('../../_chunks/dep-8f3deb2a.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var DraggerProgress = function DraggerProgress(props) {
  var file = props.file,
      onUpload = props.onUpload,
      onRemove = props.onRemove,
      display = props.display,
      onTrigger = props.onTrigger;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var reUpload = function reUpload(e) {
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      e: e,
      file: file,
      index: 0
    });
    onTrigger === null || onTrigger === void 0 ? void 0 : onTrigger();
  };

  var showResultOperate = React__default["default"].useMemo(function () {
    return ["success", "fail"].includes(file === null || file === void 0 ? void 0 : file.status);
  }, [file]);

  var handleRemove = function handleRemove(e) {
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      e: e,
      file: file,
      index: 0
    });
  };

  var renderUploading = React.useCallback(function () {
    if ((file === null || file === void 0 ? void 0 : file.status) === "fail") {
      return /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ErrorCircleFilledIcon, null);
    }

    if ((file === null || file === void 0 ? void 0 : file.status) === "progress") {
      return /* @__PURE__ */React__default["default"].createElement("div", {
        className: "t-upload__single-progress"
      }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
        loading: true,
        size: "small"
      }), /* @__PURE__ */React__default["default"].createElement("span", {
        className: "t-upload__single-percent"
      }, Math.min((file === null || file === void 0 ? void 0 : file.percent) || 0, 99), "%"));
    }
  }, [file]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-progress")
  }, display === "image" && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "t-upload__dragger-img-wrap"
  }, file && /* @__PURE__ */React__default["default"].createElement("img", {
    src: (file === null || file === void 0 ? void 0 : file.url) || "default.png",
    alt: ""
  })), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-progress-info")
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-text")
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-upload__single-name")
  }, upload_util.abridgeName(file === null || file === void 0 ? void 0 : file.name)), (file === null || file === void 0 ? void 0 : file.status) !== "success" && renderUploading(), (file === null || file === void 0 ? void 0 : file.status) === "success" && /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CheckCircleFilledIcon, null)), /* @__PURE__ */React__default["default"].createElement("small", {
    className: "".concat(classPrefix, "-size-s")
  }, "\u6587\u4EF6\u5927\u5C0F\uFF1A", upload_util.returnFileSize(file === null || file === void 0 ? void 0 : file.size)), /* @__PURE__ */React__default["default"].createElement("small", {
    className: "".concat(classPrefix, "-size-s")
  }, "\u4E0A\u4F20\u65E5\u671F\uFF1A", upload_util.getCurrentDate()), !["success", "fail"].includes(file === null || file === void 0 ? void 0 : file.status) && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-btns")
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    theme: "primary",
    variant: "text",
    className: "".concat(classPrefix, "-upload__dragger-progress-cancel"),
    onClick: handleRemove
  }, "\u53D6\u6D88\u4E0A\u4F20"), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    theme: "primary",
    variant: "text",
    onClick: onUpload
  }, "\u70B9\u51FB\u4E0A\u4F20")), showResultOperate && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "t-upload__dragger-btns"
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    theme: "primary",
    variant: "text",
    className: "t-upload__dragger-progress-cancel",
    onClick: reUpload
  }, "\u91CD\u65B0\u4E0A\u4F20"), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    theme: "primary",
    variant: "text",
    onClick: handleRemove
  }, "\u5220\u9664"))));
};

exports["default"] = DraggerProgress;
