// tdesign-react/cjs/upload/themes/image-card.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-fa36e79f.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var loading_index = require('../../loading/index.js');
var dialog_index = require('../../dialog/index.js');
var _util_useConfig = require('../../_util/useConfig.js');
var upload_util = require('../util.js');
var upload_booleanRender = require('../boolean-render.js');
require('../../_chunks/dep-045b87c1.js');
require('../../loading/loading.js');
require('../../_chunks/dep-449daca3.js');
require('classnames');
require('../../_util/dom.js');
require('raf');
require('../../_chunks/dep-0705da20.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_util/easing.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/gradient.js');
require('../../config-provider/ConfigContext.js');
require('../../locale/zh_CN.js');
require('../../loading/plugin.js');
require('../../dialog/Dialog.js');
require('../../_chunks/dep-e03ca79d.js');
require('../../locale/LocalReceiver.js');
require('../../button/index.js');
require('../../button/Button.js');
require('../../_util/noop.js');
require('../../_util/useRipple.js');
require('../../_util/setStyle.js');
require('../../dialog/RenderDialog.js');
require('react-transition-group');
require('../../_util/useSetState.js');
require('../../dialog/plugin.js');
require('../../_chunks/dep-8f3deb2a.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var ImageCard = function ImageCard(props) {
  var files = props.files,
      _props$multiple = props.multiple,
      multiple = _props$multiple === void 0 ? false : _props$multiple,
      _props$max = props.max,
      max = _props$max === void 0 ? 0 : _props$max,
      onRemove = props.onRemove,
      showUploadProgress = props.showUploadProgress;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      showImg = _useState2[0],
      setShowImg = _useState2[1];

  var _useState3 = React.useState(),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      imgURL = _useState4[0],
      setImgURL = _useState4[1];

  var preview = function preview(file) {
    setShowImg(true);
    setImgURL(file.url);
  };

  var showTrigger = React__default["default"].useMemo(function () {
    if (multiple) {
      return !max || files.length < max;
    }

    return !(files && files[0]);
  }, [files, max, multiple]);
  return /* @__PURE__ */React__default["default"].createElement(React.Fragment, null, /* @__PURE__ */React__default["default"].createElement(dialog_index.Dialog, {
    visible: showImg,
    showOverlay: true,
    width: "auto",
    top: "10%",
    className: "".concat(classPrefix, "-upload__dialog"),
    footer: false,
    header: false,
    onClose: function onClose() {
      setShowImg(false);
      setImgURL(null);
    }
  }, /* @__PURE__ */React__default["default"].createElement("p", {
    className: "".concat(classPrefix, "-dialog__dialog-body-img-box")
  }, /* @__PURE__ */React__default["default"].createElement("img", {
    src: imgURL,
    alt: ""
  }))), /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-upload__card")
  }, /* @__PURE__ */React__default["default"].createElement(upload_booleanRender["default"], {
    boolExpression: !!files
  }, files.map(function (file, index) {
    return /* @__PURE__ */React__default["default"].createElement("li", {
      className: "".concat(classPrefix, "-upload__card-item ").concat(classPrefix, "-is--background"),
      key: index
    }, /* @__PURE__ */React__default["default"].createElement(upload_booleanRender["default"], {
      falseRender: function falseRender() {
        return /* @__PURE__ */React__default["default"].createElement("div", {
          className: "".concat(classPrefix, "-upload__card-content ").concat(classPrefix, "-upload__card-box")
        }, /* @__PURE__ */React__default["default"].createElement("img", {
          className: "".concat(classPrefix, "-upload__card-image"),
          src: file.url
        }), /* @__PURE__ */React__default["default"].createElement("div", {
          className: "".concat(classPrefix, "-upload__card-mask")
        }, /* @__PURE__ */React__default["default"].createElement("span", {
          className: "".concat(classPrefix, "-upload__card-mask-item"),
          onClick: function onClick(e) {
            return e.stopPropagation();
          }
        }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.BrowseIcon, {
          onClick: function onClick() {
            preview(file);
          }
        })), /* @__PURE__ */React__default["default"].createElement("span", {
          className: "".concat(classPrefix, "-upload__card-mask-item-divider")
        }), /* @__PURE__ */React__default["default"].createElement("span", {
          className: "".concat(classPrefix, "-upload__card-mask-item"),
          onClick: function onClick(e) {
            return e.stopPropagation();
          }
        }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.DeleteIcon, {
          onClick: function onClick(e) {
            onRemove === null || onRemove === void 0 ? void 0 : onRemove({
              e: e,
              file: file,
              index: index
            });
          }
        }))));
      },
      boolExpression: showUploadProgress && !upload_util.finishUpload(file.status)
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__card-container ").concat(classPrefix, "-upload__card-box")
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
      loading: true,
      size: "medium"
    }), /* @__PURE__ */React__default["default"].createElement("p", null, "\u4E0A\u4F20\u4E2D ", Math.min(files[0].percent, 99), "%"))));
  })), /* @__PURE__ */React__default["default"].createElement(upload_booleanRender["default"], {
    boolExpression: showTrigger
  }, /* @__PURE__ */React__default["default"].createElement("li", {
    className: "".concat(classPrefix, "-upload__card-item ").concat(classPrefix, "-is--background"),
    onClick: props.onTrigger
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-upload__card-container ").concat(classPrefix, "-upload__card-box")
  }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.AddIcon, null), /* @__PURE__ */React__default["default"].createElement("p", {
    className: "".concat(classPrefix, "-size-s")
  }, "\u70B9\u51FB\u4E0A\u4F20\u56FE\u7247"))))));
};

exports["default"] = ImageCard;
