// tdesign-react/cjs/upload/themes/single-file.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var loading_index = require('../../loading/index.js');
var _util_useConfig = require('../../_util/useConfig.js');
require('../../loading/loading.js');
require('../../_chunks/dep-449daca3.js');
require('../../_chunks/dep-fa36e79f.js');
require('../../_chunks/dep-045b87c1.js');
require('../../_util/dom.js');
require('raf');
require('../../_chunks/dep-0705da20.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_util/easing.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/gradient.js');
require('../../config-provider/ConfigContext.js');
require('../../locale/zh_CN.js');
require('../../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var SingleFile = function SingleFile(props) {
  var _props$display = props.display,
      display = _props$display === void 0 ? "file" : _props$display,
      onRemove = props.onRemove,
      file = props.file,
      showUploadProgress = props.showUploadProgress;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var fileClass = classNames__default["default"]("".concat(classPrefix, "-upload__single"), "".concat(classPrefix, "-upload__single-").concat(display));
  var showProgress = showUploadProgress && file && file.status !== "success";
  var inputName = file === null || file === void 0 ? void 0 : file.name;
  var handleRemove = React.useCallback(function (e) {
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      e: e,
      file: file,
      index: 0
    });
  }, [file, onRemove]);
  var renderProgress = React.useCallback(function () {
    if ((file === null || file === void 0 ? void 0 : file.status) === "fail") {
      return /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ErrorCircleFilledIcon, null);
    }

    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__single-progress")
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
      loading: true,
      size: "small"
    }), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__single-percent")
    }, Math.min((file === null || file === void 0 ? void 0 : file.percent) || 0, 99), "%"));
  }, [classPrefix, file]);
  var renderFilePreviewAsText = React.useCallback(function () {
    if (!inputName) return;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__single-display-text ").concat(classPrefix, "-upload__display-text--margin")
    }, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__single-name")
    }, inputName), showProgress ? renderProgress() : /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CloseCircleFilledIcon, {
      className: "".concat(classPrefix, "-upload__icon-delete"),
      onClick: handleRemove
    }));
  }, [classPrefix, inputName, renderProgress, showProgress, handleRemove]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: fileClass
  }, props.children, display === "file" && renderFilePreviewAsText());
};

exports["default"] = SingleFile;
