// tdesign-react/lib/input-number/StepHandler.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-9332bbea.js';
import React from 'react';
import classNames from 'classnames';
import { ChevronDownIcon, RemoveIcon, ChevronUpIcon, AddIcon } from 'tdesign-icons-react';
import useCommonClassName from '../_util/useCommonClassName.js';
import { Button } from '../button/index.js';
import '../_util/useConfig.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../button/Button.js';
import '../_chunks/dep-27d690ce.js';
import '../_util/noop.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../loading/index.js';
import '../loading/loading.js';
import '../_chunks/dep-93971e86.js';
import '../_chunks/dep-beac373a.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-f3f6eedc.js';
import '../_chunks/dep-c6576f8f.js';
import '../_chunks/dep-156c830d.js';
import '../_chunks/dep-04305406.js';
import '../_chunks/dep-6ca758b4.js';
import '../_util/easing.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../loading/plugin.js';

function StepHandler(props) {
  var prefixClassName = props.prefixClassName,
      theme = props.theme,
      onStep = props.onStep,
      disabledDecrease = props.disabledDecrease,
      disabledIncrease = props.disabledIncrease;
  var commonClassNames = useCommonClassName();
  var decreaseIcon = theme === "column" ? /* @__PURE__ */React.createElement(ChevronDownIcon, null) : /* @__PURE__ */React.createElement(RemoveIcon, null);
  var increaseIcon = theme === "column" ? /* @__PURE__ */React.createElement(ChevronUpIcon, null) : /* @__PURE__ */React.createElement(AddIcon, null);

  var onStepDecrease = function onStepDecrease(e) {
    return disabledDecrease || onStep({
      type: "reduce",
      e: e
    });
  };

  var onStepIncrease = function onStepIncrease(e) {
    return disabledIncrease || onStep({
      type: "add",
      e: e
    });
  };

  return /* @__PURE__ */React.createElement(React.Fragment, null, /* @__PURE__ */React.createElement(Button, {
    variant: "outline",
    className: classNames("".concat(prefixClassName, "__decrease"), _defineProperty({}, commonClassNames.STATUS.disabled, disabledDecrease)),
    onClick: onStepDecrease,
    icon: decreaseIcon,
    shape: "square"
  }), /* @__PURE__ */React.createElement(Button, {
    variant: "outline",
    className: classNames("".concat(prefixClassName, "__increase"), _defineProperty({}, commonClassNames.STATUS.disabled, disabledIncrease)),
    onClick: onStepIncrease,
    icon: increaseIcon,
    shape: "square"
  }));
}

export { StepHandler as default };
