// tdesign-react/esm/upload/themes/dragger-progress.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import React, { useCallback } from 'react';
import { ErrorCircleFilledIcon, CheckCircleFilledIcon } from 'tdesign-icons-react';
import { Button } from '../../button/index.js';
import { Loading } from '../../loading/index.js';
import useConfig from '../../_util/useConfig.js';
import { abridgeName, returnFileSize, getCurrentDate } from '../util.js';
import '../../button/Button.js';
import '../../_chunks/dep-9332bbea.js';
import '../../_chunks/dep-27d690ce.js';
import 'classnames';
import '../../_util/noop.js';
import '../../_util/useRipple.js';
import '../../_util/setStyle.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../loading/loading.js';
import '../../_chunks/dep-93971e86.js';
import '../../_chunks/dep-beac373a.js';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-3853103d.js';
import '../../_chunks/dep-ab1e2688.js';
import '../../_chunks/dep-aa2bff1b.js';
import '../../_chunks/dep-156c830d.js';
import '../../_chunks/dep-897ce04d.js';
import '../../_chunks/dep-2eff142e.js';
import '../../_chunks/dep-312619d6.js';
import '../../_util/easing.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../loading/style/index.js';
import '../../loading/type.js';
import '../../loading/plugin.js';
import '../../button/style/index.js';
import '../../button/type.js';
import '../../_chunks/dep-124f91c5.js';

var DraggerProgress = function DraggerProgress(props) {
  var file = props.file,
      onUpload = props.onUpload,
      onRemove = props.onRemove,
      display = props.display,
      onTrigger = props.onTrigger;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var reUpload = function reUpload(e) {
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      e: e,
      file: file,
      index: 0
    });
    onTrigger === null || onTrigger === void 0 ? void 0 : onTrigger();
  };

  var showResultOperate = React.useMemo(function () {
    return ["success", "fail"].includes(file === null || file === void 0 ? void 0 : file.status);
  }, [file]);

  var handleRemove = function handleRemove(e) {
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      e: e,
      file: file,
      index: 0
    });
  };

  var renderUploading = useCallback(function () {
    if ((file === null || file === void 0 ? void 0 : file.status) === "fail") {
      return /* @__PURE__ */React.createElement(ErrorCircleFilledIcon, null);
    }

    if ((file === null || file === void 0 ? void 0 : file.status) === "progress") {
      return /* @__PURE__ */React.createElement("div", {
        className: "t-upload__single-progress"
      }, /* @__PURE__ */React.createElement(Loading, {
        loading: true,
        size: "small"
      }), /* @__PURE__ */React.createElement("span", {
        className: "t-upload__single-percent"
      }, Math.min((file === null || file === void 0 ? void 0 : file.percent) || 0, 99), "%"));
    }
  }, [file]);
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-progress")
  }, display === "image" && /* @__PURE__ */React.createElement("div", {
    className: "t-upload__dragger-img-wrap"
  }, file && /* @__PURE__ */React.createElement("img", {
    src: (file === null || file === void 0 ? void 0 : file.url) || "default.png",
    alt: ""
  })), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-progress-info")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-text")
  }, /* @__PURE__ */React.createElement("span", {
    className: "".concat(classPrefix, "-upload__single-name")
  }, abridgeName(file === null || file === void 0 ? void 0 : file.name)), (file === null || file === void 0 ? void 0 : file.status) !== "success" && renderUploading(), (file === null || file === void 0 ? void 0 : file.status) === "success" && /* @__PURE__ */React.createElement(CheckCircleFilledIcon, null)), /* @__PURE__ */React.createElement("small", {
    className: "".concat(classPrefix, "-size-s")
  }, "\u6587\u4EF6\u5927\u5C0F\uFF1A", returnFileSize(file === null || file === void 0 ? void 0 : file.size)), /* @__PURE__ */React.createElement("small", {
    className: "".concat(classPrefix, "-size-s")
  }, "\u4E0A\u4F20\u65E5\u671F\uFF1A", getCurrentDate()), !["success", "fail"].includes(file === null || file === void 0 ? void 0 : file.status) && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-upload__dragger-btns")
  }, /* @__PURE__ */React.createElement(Button, {
    theme: "primary",
    variant: "text",
    className: "".concat(classPrefix, "-upload__dragger-progress-cancel"),
    onClick: handleRemove
  }, "\u53D6\u6D88\u4E0A\u4F20"), /* @__PURE__ */React.createElement(Button, {
    theme: "primary",
    variant: "text",
    onClick: onUpload
  }, "\u70B9\u51FB\u4E0A\u4F20")), showResultOperate && /* @__PURE__ */React.createElement("div", {
    className: "t-upload__dragger-btns"
  }, /* @__PURE__ */React.createElement(Button, {
    theme: "primary",
    variant: "text",
    className: "t-upload__dragger-progress-cancel",
    onClick: reUpload
  }, "\u91CD\u65B0\u4E0A\u4F20"), /* @__PURE__ */React.createElement(Button, {
    theme: "primary",
    variant: "text",
    onClick: handleRemove
  }, "\u5220\u9664"))));
};

export { DraggerProgress as default };
