// tdesign-react/cjs/time-picker/panel/TimePickerRangePanel.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-449daca3.js');
var slicedToArray = require('../../_chunks/dep-fa36e79f.js');
var React = require('react');
var classNames = require('classnames');
var dayjs = require('dayjs');
var timePicker_panel_SinglePanel = require('./SinglePanel.js');
var _util_useConfig = require('../../_util/useConfig.js');
var button_index = require('../../button/index.js');
var timePicker_consts = require('../consts.js');
require('../../_chunks/dep-045b87c1.js');
require('../../_chunks/dep-228a3548.js');
require('../../_chunks/dep-735b841a.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-8bb9b7ba.js');
require('../../_chunks/dep-9c74d22d.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_chunks/dep-99d6d655.js');
require('../../_chunks/dep-7a0cfc4c.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-bc486aab.js');
require('../../_chunks/dep-c7adf4af.js');
require('../../_chunks/dep-942a7aba.js');
require('../../_chunks/dep-5e46865d.js');
require('../../_chunks/dep-059e74d9.js');
require('../../_chunks/dep-6681bc7b.js');
require('../../_chunks/dep-f445e9bb.js');
require('../../_chunks/dep-1bafebb1.js');
require('../../_chunks/dep-72511bfe.js');
require('../../_chunks/dep-7d00ab60.js');
require('../../_util/noop.js');
require('../interfaces.js');
require('../../config-provider/ConfigContext.js');
require('../../locale/zh_CN.js');
require('../../button/Button.js');
require('../../_chunks/dep-e03ca79d.js');
require('../../_util/useRipple.js');
require('../../_util/setStyle.js');
require('../../loading/index.js');
require('../../loading/loading.js');
require('../../_util/dom.js');
require('raf');
require('../../_chunks/dep-0705da20.js');
require('../../_util/easing.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/gradient.js');
require('../../loading/plugin.js');
require('../../locale/LocalReceiver.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var TimePickerPanel = function TimePickerPanel(props) {
  var _props$value = props.value,
      value = _props$value === void 0 ? [] : _props$value,
      onChange = props.onChange,
      isFooterDisplay = props.isFooterDisplay,
      handleConfirmClick = props.handleConfirmClick,
      _props$steps = props.steps,
      steps = _props$steps === void 0 ? timePicker_consts.DEFAULT_STEPS : _props$steps,
      _props$format = props.format,
      format = _props$format === void 0 ? timePicker_consts.DEFAULT_FORMAT : _props$format;
  var TEXT_CONFIG = timePicker_consts.useTimePickerTextConfig();

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _ref = value || [],
      _ref2 = slicedToArray._slicedToArray(_ref, 2),
      startTime = _ref2[0],
      endTime = _ref2[1];

  var panelClassName = "".concat(classPrefix, "-time-picker__panel");

  var handlePanelValueChange = function handlePanelValueChange(value2, index) {
    if (index === 0) {
      onChange([value2, endTime || value2]);
    } else {
      var calStartTime = startTime || dayjs__default["default"]().hour(0).minute(0).second(0).format(format);
      onChange([calStartTime, value2]);
    }
  };

  var defaultValue = React.useMemo(function () {
    if (value && value.length === 0) {
      return [dayjs__default["default"]().format(format), dayjs__default["default"]().format(format)];
    }

    return value;
  }, [value, format]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](panelClassName, "".concat(panelClassName, "-section"))
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(panelClassName, "-section-body")
  }, /* @__PURE__ */React__default["default"].createElement(timePicker_panel_SinglePanel["default"], _objectSpread(_objectSpread({}, props), {}, {
    steps: steps,
    format: format,
    value: startTime,
    onChange: function onChange(v) {
      return handlePanelValueChange(v, 0);
    }
  })), /* @__PURE__ */React__default["default"].createElement(timePicker_panel_SinglePanel["default"], _objectSpread(_objectSpread({}, props), {}, {
    steps: steps,
    format: format,
    value: endTime,
    onChange: function onChange(v) {
      return handlePanelValueChange(v, 1);
    }
  }))), isFooterDisplay ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(panelClassName, "-section-footer")
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    theme: "primary",
    variant: "base",
    onClick: function onClick() {
      handleConfirmClick(defaultValue);
    }
  }, TEXT_CONFIG.confirm)) : null);
};

exports["default"] = TimePickerPanel;
