// tdesign-react/es/tree-select/TreeSelectSuffix.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import React, { useMemo } from 'react';
import classNames from 'classnames';
import { i as isEmpty_1 } from '../_chunks/dep-a8d92bd9.js';
import { CloseCircleFilledIcon } from 'tdesign-icons-react';
import useConfig from '../_util/useConfig.js';
import FakeArrow from '../common/FakeArrow.js';
import { Loading } from '../loading/index.js';
import '../_chunks/dep-ab1e2688.js';
import '../_chunks/dep-39f90ded.js';
import '../_chunks/dep-aa2bff1b.js';
import '../_chunks/dep-156c830d.js';
import '../_chunks/dep-14692160.js';
import '../_chunks/dep-312619d6.js';
import '../_chunks/dep-55e3e3e3.js';
import '../_chunks/dep-7daf07a9.js';
import '../_chunks/dep-e2a0e7b2.js';
import '../_chunks/dep-5ace74d6.js';
import '../_chunks/dep-25536592.js';
import '../_chunks/dep-a764e3bd.js';
import '../_chunks/dep-d098d908.js';
import '../_chunks/dep-8309e373.js';
import '../_chunks/dep-f5f7d3f4.js';
import '../_chunks/dep-897ce04d.js';
import '../_chunks/dep-2eff142e.js';
import '../_chunks/dep-b8263a1c.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../_chunks/dep-9332bbea.js';
import '../loading/loading.js';
import '../_chunks/dep-93971e86.js';
import '../_chunks/dep-beac373a.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3853103d.js';
import '../_util/easing.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import './style/css.js';
import '../loading/type.js';
import '../loading/plugin.js';

function TreeSelectSuffix(props) {
  var visible = props.visible,
      isHover = props.isHover,
      handleClear = props.handleClear,
      showLoading = props.showLoading,
      clearable = props.clearable,
      multiple = props.multiple,
      value = props.value,
      disabled = props.disabled,
      size = props.size;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var showArrow = useMemo(function () {
    return !clearable || !isHover || disabled || !multiple && !value && value !== 0 || multiple && Array.isArray(value) && isEmpty_1(value);
  }, [clearable, isHover, disabled, multiple, value]);
  var showClose = useMemo(function () {
    return clearable && isHover && !disabled && (!multiple && (!!value || value === 0) || multiple && !isEmpty_1(value));
  }, [clearable, isHover, disabled, multiple, value]);
  return /* @__PURE__ */React.createElement(React.Fragment, null, showArrow && !showLoading && /* @__PURE__ */React.createElement(FakeArrow, {
    overlayClassName: "".concat(classPrefix, "-select__right-icon"),
    isActive: visible,
    disabled: disabled
  }), showClose && !showLoading && /* @__PURE__ */React.createElement(CloseCircleFilledIcon, {
    className: "".concat(classPrefix, "-select__right-icon ").concat(classPrefix, "-select__right-icon-clear"),
    size: size,
    onClick: handleClear
  }), showLoading && /* @__PURE__ */React.createElement(Loading, {
    loading: true,
    className: classNames("".concat(classPrefix, "-select__right-icon"), "".concat(classPrefix, "-select__active-icon")),
    size: "small"
  }));
}

export { TreeSelectSuffix as default };
