// tdesign-react/cjs/upload/dragger.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-fa36e79f.js');
var React = require('react');
var classNames = require('classnames');
var _util_useConfig = require('../_util/useConfig.js');
var upload_themes_draggerProgress = require('./themes/dragger-progress.js');
require('../_chunks/dep-045b87c1.js');
require('../config-provider/ConfigContext.js');
require('../locale/zh_CN.js');
require('tdesign-icons-react');
require('../button/index.js');
require('../button/Button.js');
require('../_chunks/dep-449daca3.js');
require('../_chunks/dep-e03ca79d.js');
require('../_util/noop.js');
require('../_util/useRipple.js');
require('../_util/setStyle.js');
require('../loading/index.js');
require('../loading/loading.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-0705da20.js');
require('../_chunks/dep-0e3f1753.js');
require('../_chunks/dep-a7666a95.js');
require('../_chunks/dep-9125a75e.js');
require('../_chunks/dep-48870ae5.js');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../loading/plugin.js');
require('./util.js');
require('../_chunks/dep-8f3deb2a.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Dragger = function Dragger(props) {
  var file = props.file,
      display = props.display,
      _onUpload = props.onUpload,
      onRemove = props.onRemove,
      customDraggerRender = props.customDraggerRender;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      dragActive = _useState2[0],
      setDragActive = _useState2[1];

  var target = React__default["default"].useRef();
  var classes = classNames__default["default"]("".concat(classPrefix, "-upload__dragger"), !file ? "".concat(classPrefix, "-upload__dragger-center") : "", (file === null || file === void 0 ? void 0 : file.status) === "fail" ? "".concat(classPrefix, "-upload__dragger-error") : "");
  var defaultDragElement = React__default["default"].useMemo(function () {
    var unActiveElement = /* @__PURE__ */React__default["default"].createElement("div", null, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload--highlight")
    }, "\u70B9\u51FB\u4E0A\u4F20"), /* @__PURE__ */React__default["default"].createElement("span", null, "\xA0\xA0/\xA0\xA0\u62D6\u62FD\u5230\u6B64\u533A\u57DF"));
    var activeElement = /* @__PURE__ */React__default["default"].createElement("div", null, "\u91CA\u653E\u9F20\u6807");
    return dragActive ? activeElement : unActiveElement;
  }, [classPrefix, dragActive]);
  var dragElement = React__default["default"].useMemo(function () {
    var content;

    if (file && display !== "custom") {
      content = /* @__PURE__ */React__default["default"].createElement(upload_themes_draggerProgress["default"], {
        onRemove: onRemove,
        display: display,
        onTrigger: props.onTrigger,
        file: file,
        onUpload: function onUpload() {
          _onUpload === null || _onUpload === void 0 ? void 0 : _onUpload(file);
        }
      });
    } else {
      content = /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-upload__trigger"),
        onClick: props.onTrigger
      }, (customDraggerRender === null || customDraggerRender === void 0 ? void 0 : customDraggerRender({
        dragActive: dragActive
      })) || defaultDragElement);
    }

    return content;
  }, [classPrefix, defaultDragElement, display, dragActive, file, onRemove, _onUpload, props, customDraggerRender]);

  var handleDrop = function handleDrop(event) {
    var _props$onChange, _props$onDragleave;

    event.preventDefault();
    (_props$onChange = props.onChange) === null || _props$onChange === void 0 ? void 0 : _props$onChange.call(props, event.dataTransfer.files);
    (_props$onDragleave = props.onDragleave) === null || _props$onDragleave === void 0 ? void 0 : _props$onDragleave.call(props, event);
    setDragActive(false);
  };

  var handleDragenter = function handleDragenter(event) {
    var _props$onDragenter;

    event.preventDefault();
    (_props$onDragenter = props.onDragenter) === null || _props$onDragenter === void 0 ? void 0 : _props$onDragenter.call(props, event);
    setDragActive(true);
  };

  var handleDragleave = function handleDragleave(event) {
    var _props$onDragleave2;

    if (!target.current) return;
    event.preventDefault();
    (_props$onDragleave2 = props.onDragleave) === null || _props$onDragleave2 === void 0 ? void 0 : _props$onDragleave2.call(props, event);
    setDragActive(false);
  };

  var handleDragover = function handleDragover(event) {
    event.preventDefault();
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: target,
    className: classes,
    onDrop: handleDrop,
    onDragEnter: handleDragenter,
    onDragOver: handleDragover,
    onDragLeave: handleDragleave
  }, dragElement);
};

exports["default"] = Dragger;
