// tdesign-react/cjs/cascader/utils/item.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-8f3deb2a.js');
var defineProperty = require('../../_chunks/dep-449daca3.js');
require('../../_chunks/dep-045b87c1.js');

function getLabelIsEllipsis(node, size) {
  var sizeMap = {
    small: 11,
    medium: 9,
    large: 8
  };
  return sizeMap[size] < node.label.length;
}
function getNodeStatusClass(node, CLASSNAMES, cascaderContext) {
  var _ref;

  var checkStrictly = cascaderContext.checkStrictly,
      multiple = cascaderContext.multiple,
      value = cascaderContext.value,
      max = cascaderContext.max;
  var expandedActive = !checkStrictly && node.expanded && (multiple ? !node.isLeaf() : true);
  var isLeaf = node.isLeaf();
  var isDisabled = node.disabled || multiple && value.length >= max && max !== 0;
  var isSelected = node.checked || multiple && !checkStrictly && node.expanded && !isLeaf;
  return [(_ref = {}, defineProperty._defineProperty(_ref, CLASSNAMES.STATUS.selected, !isDisabled && isSelected), defineProperty._defineProperty(_ref, CLASSNAMES.STATUS.expanded, !isDisabled && expandedActive), defineProperty._defineProperty(_ref, CLASSNAMES.STATUS.disabled, isDisabled), _ref)];
}
function getCascaderItemClass(prefix, node, CLASSNAMES, cascaderContext) {
  var _ref2;

  var size = cascaderContext.size;
  return ["".concat(prefix, "-cascader__item")].concat(toConsumableArray._toConsumableArray(getNodeStatusClass(node, CLASSNAMES, cascaderContext)), [(_ref2 = {}, defineProperty._defineProperty(_ref2, "".concat(prefix, "-cascader__item--with-icon"), node.children), defineProperty._defineProperty(_ref2, "".concat(prefix, "-cascader__item--leaf"), node.isLeaf()), defineProperty._defineProperty(_ref2, CLASSNAMES.SIZE[size], size), _ref2)]);
}
function getCascaderItemIconClass(prefix, node, CLASSNAMES, cascaderContext) {
  return ["".concat(prefix, "-cascader__item-icon")].concat(toConsumableArray._toConsumableArray(getNodeStatusClass(node, CLASSNAMES, cascaderContext)));
}

exports.getCascaderItemClass = getCascaderItemClass;
exports.getCascaderItemIconClass = getCascaderItemIconClass;
exports.getLabelIsEllipsis = getLabelIsEllipsis;
exports.getNodeStatusClass = getNodeStatusClass;
