// tdesign-react/cjs/table/primary/SorterButton.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-449daca3.js');
var slicedToArray = require('../../_chunks/dep-fa36e79f.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var tooltip_index = require('../../tooltip/index.js');
var configProvider_ConfigContext = require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-045b87c1.js');
require('../../tooltip/Tooltip.js');
require('../../_chunks/dep-e03ca79d.js');
require('../../_chunks/dep-a7666a95.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../../_util/useDefault.js');
require('../../_util/noop.js');
require('../../_util/useConfig.js');
require('../../locale/zh_CN.js');
require('../../_util/composeRefs.js');
require('../../_util/usePrevious.js');
require('../../popup/Portal.js');
require('react-dom');
require('../../popup/hooks/useTriggerProps.js');
require('../../_util/useClickOutside.js');
require('../../popup/hooks/usePopupCssTransition.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var SortTypeEnum = {
  desc: "desc",
  asc: "asc",
  all: "all"
};
var sortTypeOrder = [SortTypeEnum.desc, SortTypeEnum.asc, SortTypeEnum.all];
var tooltips = {
  desc: "\u70B9\u51FB\u964D\u5E8F",
  asc: "\u70B9\u51FB\u5347\u5E8F",
  all: "\u70B9\u51FB\u53D6\u6D88\u6392\u5E8F"
};

var SorterButton = function SorterButton(props) {
  var currentColumn = props.column,
      currentSort = props.singleSort,
      onChange = props.onChange;
  var _currentColumn$sortTy = currentColumn.sortType,
      sortType = _currentColumn$sortTy === void 0 ? SortTypeEnum.all : _currentColumn$sortTy;

  var _useContext = React.useContext(configProvider_ConfigContext["default"]),
      classPrefix = _useContext.classPrefix;

  var _useState = React.useState(SortTypeEnum.all),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      currentSortType = _useState2[0],
      setCurrentSortType = _useState2[1];

  var isAllSortType = sortType === SortTypeEnum.all;
  var nextSortType = getNextSortType(currentSortType);

  function getNextSortType(currentSortType2) {
    if (isAllSortType) {
      var currentIndex = sortTypeOrder.indexOf(currentSortType2);
      var nextIndex = currentIndex === sortTypeOrder.length - 1 ? 0 : currentIndex + 1;
      return sortTypeOrder[nextIndex];
    }

    return currentSortType2 === SortTypeEnum.all ? sortType : SortTypeEnum.all;
  }

  function renderIcon(direction, className) {
    var style = {};

    if (direction === SortTypeEnum.asc) {
      style = _objectSpread({
        transform: "rotate(-180deg)",
        top: "-1px"
      }, style);
    } else {
      style.bottom = "-1px";
    }

    var sortClassName = classNames__default["default"](["".concat(classPrefix, "-table__sort-icon"), className, "".concat(classPrefix, "-table__sort-").concat(direction)]);
    return /* @__PURE__ */React__default["default"].createElement("div", {
      key: direction,
      style: style,
      className: sortClassName
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ChevronDownIcon, {
      size: "16px"
    }));
  }

  function renderSortIcons() {
    var allowSortTypes = [];

    if (isAllSortType) {
      allowSortTypes.push(SortTypeEnum.asc, SortTypeEnum.desc);
    } else {
      allowSortTypes.push(sortType);
    }

    var sortIcons = allowSortTypes.map(function (direction) {
      var className = direction === currentSortType ? "".concat(classPrefix, "-table__sort-icon--active") : "".concat(classPrefix, "-icon-sort-default");
      return renderIcon(direction, className);
    });
    return sortIcons.length > 1 ? /* @__PURE__ */React__default["default"].createElement("div", {
      style: {
        lineHeight: 0
      }
    }, sortIcons) : sortIcons;
  }

  function handleSort() {
    var currentSortTypeNew = nextSortType;
    setCurrentSortType(currentSortTypeNew);
    onChange(currentSort, currentSortTypeNew, currentColumn);
  }

  React.useEffect(function () {
    var currentSortType2 = SortTypeEnum.all;

    if (currentSort) {
      currentSortType2 = currentSort.descending ? SortTypeEnum.desc : SortTypeEnum.asc;
    }

    setCurrentSortType(currentSortType2);
  }, [currentSort]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](["".concat(classPrefix, "-table__cell--sort-trigger"), defineProperty._defineProperty({}, "".concat(classPrefix, "-table__double-icons"), true)]),
    onClick: handleSort
  }, /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, {
    content: tooltips[nextSortType],
    showArrow: false
  }, renderSortIcons()));
};

exports.SortTypeEnum = SortTypeEnum;
exports["default"] = SorterButton;
