// tdesign-react/cjs/table/primary/useFilter.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-449daca3.js');
var slicedToArray = require('../../_chunks/dep-fa36e79f.js');
var React = require('react');
var table_primary_FilterButton = require('./FilterButton.js');
require('../../_chunks/dep-045b87c1.js');
require('classnames');
require('../../_chunks/dep-1bafebb1.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-735b841a.js');
require('../../_chunks/dep-0705da20.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-48870ae5.js');
require('tdesign-icons-react');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../../_util/useDefault.js');
require('../../_util/noop.js');
require('../../_util/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../locale/zh_CN.js');
require('../../_util/composeRefs.js');
require('../../_util/usePrevious.js');
require('../../popup/Portal.js');
require('react-dom');
require('../../popup/hooks/useTriggerProps.js');
require('../../_util/useClickOutside.js');
require('../../popup/hooks/usePopupCssTransition.js');
require('../../radio/index.js');
require('../../_chunks/dep-f1b75399.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../common/Check.js');
require('../../_chunks/dep-e03ca79d.js');
require('../../_util/helper.js');
require('../../_chunks/dep-3ca698ba.js');
require('../../_chunks/dep-7a0cfc4c.js');
require('../../_chunks/dep-9c74d22d.js');
require('../../_chunks/dep-bc486aab.js');
require('../../_chunks/dep-c7adf4af.js');
require('../../_util/useCommonClassName.js');
require('../../checkbox/index.js');
require('../../_chunks/dep-8027cfc6.js');
require('../../_chunks/dep-f0639562.js');
require('../../input/index.js');
require('../../input/Input.js');
require('../../input/InputGroup.js');
require('../../_util/useDefaultValue.js');

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function useFilter(props) {
  var columns = props.columns,
      filterIcon = props.filterIcon,
      filterValue = props.filterValue,
      defaultFilterValue = props.defaultFilterValue,
      onFilterChange = props.onFilterChange,
      data = props.data;
  var isControlled = typeof filterValue !== "undefined";

  var _useState = React.useState(defaultFilterValue),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      filterVal = _useState2[0],
      setFilterVal = _useState2[1];

  React.useEffect(function () {
    setFilterVal(isControlled ? filterValue : defaultFilterValue);
  }, [filterValue, defaultFilterValue, isControlled]);
  var transformedFilterData = React.useMemo(function () {
    if (isControlled) {
      return data;
    }

    if (filterVal) {
      var tmpData = data;
      Object.keys(filterVal).forEach(function (k) {
        if (typeof (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) === "string" || typeof (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) === "number") {
          tmpData = tmpData.filter(function (item) {
            var _item$k;

            return item === null || item === void 0 ? void 0 : (_item$k = item[k]) === null || _item$k === void 0 ? void 0 : _item$k.toString().includes(filterVal === null || filterVal === void 0 ? void 0 : filterVal[k].toString());
          });
        }

        if (Array.isArray(filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) && (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]).length > 0) {
          tmpData = tmpData.filter(function (item) {
            return (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]).includes(item === null || item === void 0 ? void 0 : item[k]);
          });
        }
      });
      return tmpData;
    }

    return data;
  }, [data, filterVal, isControlled]);
  var transformedFilterColumns = table_primary_FilterButton["default"]({
    columns: columns,
    onChange: onChangeFilterButton,
    innerfiltVal: isControlled ? filterValue : defaultFilterValue,
    filterIcon: filterIcon
  });

  function onChangeFilterButton(toast, col) {
    var newFilterVal = _objectSpread(_objectSpread({}, filterVal), toast);

    setFilterVal(newFilterVal);

    if (onFilterChange) {
      onFilterChange(newFilterVal, {
        col: col
      });
    }
  }

  return [transformedFilterColumns, transformedFilterData];
}

exports["default"] = useFilter;
