// tdesign-react/cjs/table/primary/useSorter.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-8f3deb2a.js');
var defineProperty = require('../../_chunks/dep-449daca3.js');
var slicedToArray = require('../../_chunks/dep-fa36e79f.js');
var React = require('react');
var get = require('../../_chunks/dep-af921283.js');
var configProvider_ConfigContext = require('../../config-provider/ConfigContext.js');
var table_primary_SorterButton = require('./SorterButton.js');
require('../../_chunks/dep-045b87c1.js');
require('../../_chunks/dep-4f605641.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-9c74d22d.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_chunks/dep-20e76bfc.js');
require('../../_chunks/dep-1bafebb1.js');
require('../../_chunks/dep-735b841a.js');
require('../../_chunks/dep-6681bc7b.js');
require('../../_chunks/dep-7a0cfc4c.js');
require('../../locale/zh_CN.js');
require('classnames');
require('tdesign-icons-react');
require('../../tooltip/index.js');
require('../../tooltip/Tooltip.js');
require('../../_chunks/dep-e03ca79d.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../../_util/useDefault.js');
require('../../_util/noop.js');
require('../../_util/useConfig.js');
require('../../_util/composeRefs.js');
require('../../_util/usePrevious.js');
require('../../popup/Portal.js');
require('react-dom');
require('../../popup/hooks/useTriggerProps.js');
require('../../_util/useClickOutside.js');
require('../../popup/hooks/usePopupCssTransition.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function useSorter(props) {
  var _useContext = React.useContext(configProvider_ConfigContext["default"]),
      classPrefix = _useContext.classPrefix;

  var columns = props.columns,
      sort = props.sort,
      defaultSort = props.defaultSort,
      multipleSort = props.multipleSort,
      onSortChange = props.onSortChange,
      data = props.data;
  var isControlled = typeof sort !== "undefined";

  var _useState = React.useState(defaultSort || []),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      innerSort = _useState2[0],
      setInnerSort = _useState2[1];

  var sorts = getSorts(innerSort, columns);
  var transformedSorterColumns = columns.map(function (column) {
    var _sorts$find;

    var title = column.title,
        sorter = column.sorter,
        sortType = column.sortType,
        colKey = column.colKey;

    if (!sorter || !table_primary_SorterButton.SortTypeEnum[sortType]) {
      return column;
    }

    var singleSort = (_sorts$find = sorts.find) === null || _sorts$find === void 0 ? void 0 : _sorts$find.call(sorts, function (sortItem) {
      return (sortItem === null || sortItem === void 0 ? void 0 : sortItem.sortBy) === colKey;
    });

    var titleNew = function titleNew() {
      return /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-table__cell--sortable")
      }, /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-table__cell--title")
      }, /* @__PURE__ */React__default["default"].createElement("div", null, title), /* @__PURE__ */React__default["default"].createElement(table_primary_SorterButton["default"], {
        column: column,
        singleSort: singleSort,
        onChange: onChangeSortButton
      })));
    };

    return _objectSpread(_objectSpread({}, column), {}, {
      title: titleNew
    });
  });
  var transformedSorterData = React.useMemo(function () {
    if (isControlled) {
      return data;
    }

    return toConsumableArray._toConsumableArray(data).sort(comparer(sorts));
  }, [data, sorts, isControlled]);

  function getSorts(innerSort2, columns2) {
    var _validSorts;

    var validSorts = [];
    var sortColumns = columns2.filter(function (_ref) {
      var sorter = _ref.sorter;
      return !!sorter;
    });
    var colKeySorterMap = {};
    sortColumns.forEach(function (_ref2) {
      var colKey = _ref2.colKey,
          sorter = _ref2.sorter;
      colKeySorterMap[colKey] = sorter;
    });
    var sortColKeys = sortColumns.map(function (_ref3) {
      var colKey = _ref3.colKey;
      return colKey;
    });

    if (Array.isArray(innerSort2)) {
      validSorts = innerSort2.filter(function (sortItem) {
        return sortColKeys.includes(sortItem === null || sortItem === void 0 ? void 0 : sortItem.sortBy);
      });
    } else if (sortColKeys.includes(innerSort2 === null || innerSort2 === void 0 ? void 0 : innerSort2.sortBy)) {
      validSorts = [innerSort2];
    }

    var validSortsWithSorter = (_validSorts = validSorts) === null || _validSorts === void 0 ? void 0 : _validSorts.map(function (sortItem) {
      return _objectSpread(_objectSpread({}, sortItem), {}, {
        sorter: colKeySorterMap[sortItem.sortBy]
      });
    });
    return validSortsWithSorter;
  }

  function onChangeSortButton(activeSort, activeSortType, activeColumn) {
    var sortOptions = {
      currentDataSource: transformedSorterData,
      col: activeColumn
    };
    var activeColKey = activeColumn.colKey;

    if (multipleSort) {
      var sortsNew = [];

      if (activeSort) {
        if (activeSortType === table_primary_SorterButton.SortTypeEnum.all) {
          sortsNew = sorts.filter(function (sortItem) {
            return sortItem !== activeSort;
          }).map(function (_ref4) {
            var sortBy = _ref4.sortBy,
                descending = _ref4.descending;
            return {
              sortBy: sortBy,
              descending: descending
            };
          });
        } else {
          sortsNew = sorts.map(function (sortItem) {
            return {
              sortBy: sortItem.sortBy,
              descending: sortItem === activeSort ? activeSortType === table_primary_SorterButton.SortTypeEnum.desc : sortItem.descending
            };
          });
        }
      } else {
        var sortInfo = {
          sortBy: activeColKey,
          descending: activeSortType === table_primary_SorterButton.SortTypeEnum.desc
        };

        if (sorts.length) {
          sortsNew = [].concat(toConsumableArray._toConsumableArray(sorts.map(function (_ref5) {
            var sortBy = _ref5.sortBy,
                descending = _ref5.descending;
            return {
              sortBy: sortBy,
              descending: descending
            };
          })), [sortInfo]);
        } else {
          sortsNew = [sortInfo];
        }
      }

      setInnerSort(sortsNew);
      onSortChange(sortsNew, sortOptions);
    } else {
      var sortNew;

      if (activeSort) {
        if (activeSortType !== table_primary_SorterButton.SortTypeEnum.all) {
          sortNew = {
            sortBy: activeSort.sortBy,
            descending: activeSortType === table_primary_SorterButton.SortTypeEnum.desc
          };
        }
      } else {
        sortNew = {
          sortBy: activeColKey,
          descending: activeSortType === table_primary_SorterButton.SortTypeEnum.desc
        };
      }

      setInnerSort(sortNew);
      onSortChange(sortNew, sortOptions);
    }
  }

  React.useEffect(function () {
    if (isControlled) {
      setInnerSort(sort);
    }
  }, [sort, isControlled]);
  return [transformedSorterColumns, transformedSorterData];
}

function comparer(sorts) {
  return function (a, b) {
    var _iterator = _createForOfIteratorHelper(sorts),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var _step$value = _step.value,
            sortBy = _step$value.sortBy,
            descending = _step$value.descending,
            sorter = _step$value.sorter;
        var firstValue = get.get_1(a, sortBy);
        var secondValue = get.get_1(b, sortBy);
        var isAsc = typeof sorter === "function" ? sorter(a, b) < 0 : firstValue < secondValue;
        var isDesc = typeof sorter === "function" ? sorter(a, b) > 0 : firstValue > secondValue;

        if (isAsc || isDesc) {
          if (!descending) return isAsc ? -1 : 1;
          if (descending) return isDesc ? -1 : 1;
        }
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }

    return 0;
  };
}

exports["default"] = useSorter;
