// tdesign-react/cjs/table/primary/useSelect.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-8f3deb2a.js');
var defineProperty = require('../../_chunks/dep-449daca3.js');
var slicedToArray = require('../../_chunks/dep-fa36e79f.js');
var React = require('react');
var classNames = require('classnames');
var isFunction = require('../../_chunks/dep-1bafebb1.js');
var configProvider_ConfigContext = require('../../config-provider/ConfigContext.js');
var checkbox_index = require('../../checkbox/index.js');
var radio_index = require('../../radio/index.js');
require('../../_chunks/dep-045b87c1.js');
require('../../_chunks/dep-0e3f1753.js');
require('../../_chunks/dep-a7666a95.js');
require('../../_chunks/dep-735b841a.js');
require('../../locale/zh_CN.js');
require('../../_chunks/dep-8027cfc6.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../common/Check.js');
require('../../_chunks/dep-e03ca79d.js');
require('../../_util/helper.js');
require('../../_chunks/dep-3ca698ba.js');
require('../../_chunks/dep-7a0cfc4c.js');
require('../../_chunks/dep-9125a75e.js');
require('../../_chunks/dep-9c74d22d.js');
require('../../_chunks/dep-48870ae5.js');
require('../../_chunks/dep-bc486aab.js');
require('../../_chunks/dep-c7adf4af.js');
require('../../_util/useConfig.js');
require('../../_util/useDefault.js');
require('../../_util/noop.js');
require('../../_chunks/dep-f0639562.js');
require('../../_chunks/dep-f1b75399.js');
require('../../_util/useCommonClassName.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var TypeEnum = {
  single: "single",
  multiple: "multiple"
};
var defaultWidth = 50;

function useSelect(props) {
  var _useContext = React.useContext(configProvider_ConfigContext["default"]),
      classPrefix = _useContext.classPrefix;

  var columns = props.columns,
      data = props.data,
      rowKey = props.rowKey,
      defaultSelectedRowKeys = props.defaultSelectedRowKeys,
      selectedRowKeys = props.selectedRowKeys,
      onSelectChange = props.onSelectChange;
  var isControlled = !!selectedRowKeys;

  var _useState = React.useState(defaultSelectedRowKeys || []),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      innerSelectedRowKeys = _useState2[0],
      setInnerSelectedRowKeys = _useState2[1];

  var selectColumn = columns.find(function (column) {
    return !!TypeEnum[column.type];
  });
  React.useEffect(function () {
    if (isControlled) {
      setInnerSelectedRowKeys(selectedRowKeys);
    }
  }, [selectedRowKeys, isControlled]);

  if (!selectColumn) {
    return [columns];
  }

  var disabledFn = getDisableFn({
    selectColumn: selectColumn
  });
  var titleCheckboxProps = getTitleCheckboxProps({
    data: data,
    disabledFn: disabledFn,
    innerSelectedRowKeys: innerSelectedRowKeys,
    selectColumn: selectColumn,
    TypeEnum: TypeEnum
  });
  var transformedColumns = columns.map(function (column) {
    var type = column.type,
        className = column.className,
        width = column.width;

    if (!TypeEnum[type]) {
      return column;
    }

    var title;
    var cell;
    var isMultiple = type === TypeEnum.multiple;

    if (isMultiple) {
      title = function title() {
        var indeterminate = titleCheckboxProps.indeterminate,
            checked = titleCheckboxProps.checked,
            disabled = titleCheckboxProps.disabled;
        return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, {
          indeterminate: indeterminate,
          checked: checked,
          disabled: disabled,
          onChange: function onChange(checked2) {
            return onTitleCheckboxChange(checked2);
          }
        });
      };

      cell = function cell(options) {
        var row = options.row;
        var currentRowKeyValue = row[rowKey];
        var checked = innerSelectedRowKeys.includes(currentRowKeyValue);
        var disabled = disabledFn(options);
        return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, {
          checked: checked,
          disabled: disabled,
          onChange: function onChange(checked2) {
            return onRowCheckboxChange(currentRowKeyValue, row, checked2);
          }
        });
      };
    } else {
      cell = function cell(options) {
        var row = options.row;
        var currentRowKeyValue = row[rowKey];
        var checked = innerSelectedRowKeys.includes(currentRowKeyValue);
        var disabled = disabledFn(options);
        return /* @__PURE__ */React__default["default"].createElement(radio_index.Radio, {
          checked: checked,
          disabled: disabled,
          onChange: function onChange() {
            return onRadioChange(currentRowKeyValue, row);
          }
        });
      };
    }

    return _objectSpread({
      width: width || defaultWidth,
      title: title,
      cell: cell,
      style: {
        padding: "10px 0 10px 24px"
      },
      className: classNames__default["default"](["".concat(classPrefix, "-table__cell--selectable"), defineProperty._defineProperty({}, "".concat(className), !!className)])
    }, column);
  });

  function getTitleCheckboxProps(_ref2) {
    var data2 = _ref2.data,
        disabledFn2 = _ref2.disabledFn,
        innerSelectedRowKeys2 = _ref2.innerSelectedRowKeys,
        selectColumn2 = _ref2.selectColumn,
        TypeEnum2 = _ref2.TypeEnum;

    if (selectColumn2.type !== TypeEnum2.multiple) {
      return {
        checked: false,
        indeterminate: false,
        disabled: false
      };
    }

    var _getRowKeysExcludeDis = getRowKeysExcludeDisabledAndIsDisabledAll({
      data: data2,
      disabledFn: disabledFn2,
      innerSelectedRowKeys: innerSelectedRowKeys2
    }),
        rowKeysExcludeDisabled = _getRowKeysExcludeDis.rowKeysExcludeDisabled,
        isDisabledAll = _getRowKeysExcludeDis.isDisabledAll;

    var innerSelectedRowKeysExcludeDisabled = innerSelectedRowKeys2.filter(function (rowKey2) {
      return rowKeysExcludeDisabled.includes(rowKey2);
    });
    var checked = innerSelectedRowKeysExcludeDisabled.length && innerSelectedRowKeysExcludeDisabled.length === rowKeysExcludeDisabled.length;
    var indeterminate = innerSelectedRowKeysExcludeDisabled.length && innerSelectedRowKeysExcludeDisabled.length < rowKeysExcludeDisabled.length;
    return {
      checked: checked,
      indeterminate: indeterminate,
      disabled: isDisabledAll
    };
  }

  function getDisableFn(_ref3) {
    var selectColumn2 = _ref3.selectColumn;
    var disabled = selectColumn2.disabled,
        checkProps = selectColumn2.checkProps;

    var disabledFn2 = function disabledFn2() {
      return false;
    };

    if (disabled && isFunction.isFunction_1(disabled)) {
      disabledFn2 = disabled;
    } else if (checkProps) {
      if (isFunction.isFunction_1(checkProps)) {
        disabledFn2 = function disabledFn2(options) {
          var _checkProps;

          return (_checkProps = checkProps(options)) === null || _checkProps === void 0 ? void 0 : _checkProps.disabled;
        };
      } else if (checkProps.disabled) {
        disabledFn2 = function disabledFn2() {
          return checkProps.disabled;
        };
      }
    }

    return disabledFn2;
  }

  function getRowKeysExcludeDisabledAndIsDisabledAll(_ref4) {
    var data2 = _ref4.data,
        disabledFn2 = _ref4.disabledFn,
        innerSelectedRowKeys2 = _ref4.innerSelectedRowKeys;
    var dataDisabled = [];
    var rowKeysExcludeDisabled = [];
    data2.forEach(function (dataItem, index) {
      var isDisabled = disabledFn2({
        rowIndex: index,
        row: dataItem
      });
      innerSelectedRowKeys2.includes(dataItem[rowKey]);

      if (!isDisabled) {
        rowKeysExcludeDisabled.push(dataItem[rowKey]);
      } else {
        dataDisabled.push(dataItem);
      }
    });
    var isDisabledAll = dataDisabled.length === data2.length;
    return {
      rowKeysExcludeDisabled: rowKeysExcludeDisabled,
      isDisabledAll: isDisabledAll
    };
  }

  function onTitleCheckboxChange(checked) {
    var selectedRowData = [];

    if (checked) {
      data.forEach(function (dataItem, index) {
        var isDisabled = disabledFn({
          rowIndex: index,
          row: dataItem
        });
        var isChecked = innerSelectedRowKeys.includes(dataItem[rowKey]);

        if (!isDisabled || isChecked) {
          selectedRowData.push(dataItem);
        }
      });
    } else {
      data.forEach(function (dataItem, index) {
        var isDisabled = disabledFn({
          rowIndex: index,
          row: dataItem
        });
        var isChecked = innerSelectedRowKeys.includes(dataItem[rowKey]);

        if (isDisabled && isChecked) {
          selectedRowData.push(dataItem);
        }
      });
    }

    var selectedRowKeysNew = selectedRowData.map(function (record) {
      return record[rowKey];
    });

    if (!isControlled) {
      setInnerSelectedRowKeys(selectedRowKeysNew);
    }

    isFunction.isFunction_1(onSelectChange) && onSelectChange(selectedRowKeysNew, {
      selectedRowData: selectedRowData
    });
  }

  function onRowCheckboxChange(currentRowKeyValue, row, checked) {
    var selectedRowKeysNew = [];

    if (checked) {
      selectedRowKeysNew = [].concat(toConsumableArray._toConsumableArray(innerSelectedRowKeys), [currentRowKeyValue]);
    } else {
      selectedRowKeysNew = innerSelectedRowKeys.filter(function (selectRowKey) {
        return selectRowKey !== currentRowKeyValue;
      });
    }

    var selectedRowData = data.filter(function (dataItem) {
      return selectedRowKeysNew.includes(dataItem[rowKey]);
    });

    if (!isControlled) {
      setInnerSelectedRowKeys(selectedRowKeysNew);
    }

    isFunction.isFunction_1(onSelectChange) && onSelectChange(selectedRowKeysNew, {
      selectedRowData: selectedRowData
    });
  }

  function onRadioChange(currentRowKeyValue, row) {
    var selectedRowKeysNew = [currentRowKeyValue];
    var selectedRowData = [row];

    if (!isControlled) {
      setInnerSelectedRowKeys(selectedRowKeysNew);
    }

    isFunction.isFunction_1(onSelectChange) && onSelectChange(selectedRowKeysNew, {
      selectedRowData: selectedRowData
    });
  }

  return [transformedColumns];
}

exports["default"] = useSelect;
