// tdesign-react/cjs/_chunks/dep-8027cfc6.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

var defineProperty = require('./dep-449daca3.js');
var React = require('react');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var common_Check = require('../common/Check.js');
var slicedToArray = require('./dep-fa36e79f.js');
var _typeof = require('./dep-a7666a95.js');
var classNames = require('classnames');
var isNumber = require('./dep-f0639562.js');
var _util_useConfig = require('../_util/useConfig.js');
var _util_useDefault = require('../_util/useDefault.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys$1(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys$1(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var getCheckboxValue = function getCheckboxValue(v) {
  switch (_typeof._typeof(v)) {
    case "number":
      return v;

    case "object":
      {
        var vs = v;
        return vs.value;
      }

    default:
      return void 0;
  }
};

function CheckboxGroup(props) {
  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var value = props.value,
      defaultValue = props.defaultValue,
      onChange = props.onChange,
      disabled = props.disabled,
      className = props.className,
      style = props.style,
      children = props.children,
      max = props.max,
      _props$options = props.options,
      options = _props$options === void 0 ? [] : _props$options;
  var intervalOptions = Array.isArray(options) && options.length > 0 ? options : React__default["default"].Children.map(children, function (child) {
    return child.props;
  });
  var optionsWithoutCheckAll = intervalOptions.filter(function (t) {
    return _typeof._typeof(t) !== "object" || !t.checkAll;
  });
  var optionsWithoutCheckAllValues = [];
  optionsWithoutCheckAll.forEach(function (v) {
    var vs = getCheckboxValue(v);
    optionsWithoutCheckAllValues.push(vs);
  });

  var _useDefault = _util_useDefault["default"](value, defaultValue, onChange),
      _useDefault2 = slicedToArray._slicedToArray(_useDefault, 2),
      internalValue = _useDefault2[0],
      setInternalValue = _useDefault2[1];

  var _useState = React.useState(max),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      localMax = _useState2[0],
      setLocalMax = _useState2[1];

  var checkedSet = React.useMemo(function () {
    if (!Array.isArray(internalValue)) return /* @__PURE__ */new Set([]);
    return new Set([].concat(internalValue));
  }, [internalValue]);
  var indeterminate = React.useMemo(function () {
    var list = Array.from(checkedSet);
    return list.length !== 0 && list.length !== optionsWithoutCheckAll.length;
  }, [checkedSet, optionsWithoutCheckAll]);
  var checkAllChecked = React.useMemo(function () {
    var list = Array.from(checkedSet);
    return list.length === optionsWithoutCheckAll.length;
  }, [checkedSet, optionsWithoutCheckAll]);
  React.useEffect(function () {
    if (!isNumber.isNumber_1(max)) return;

    if (max < checkedSet.size) {
      console.warn("[TDesign] max should be less than the length of value, change is invalid");
    } else {
      setLocalMax(max);
    }
  }, [max, checkedSet]);
  var context = {
    inject: function inject(checkProps) {
      if (typeof checkProps.checked !== "undefined") {
        return checkProps;
      }

      var checkValue = checkProps.value;
      return _objectSpread$1(_objectSpread$1({}, checkProps), {}, {
        checked: checkProps.checkAll ? checkAllChecked : checkedSet.has(checkValue),
        indeterminate: checkProps.checkAll ? indeterminate : checkProps.indeterminate,
        disabled: checkProps.disabled || disabled || checkedSet.size >= localMax && !checkedSet.has(checkValue),
        onChange: function onChange(checked, _ref) {
          var e = _ref.e;

          if (typeof checkProps.onChange === "function") {
            checkProps.onChange(checked, {
              e: e
            });
          }

          if (checkProps.checkAll) {
            checkedSet.clear();

            if (checked) {
              optionsWithoutCheckAllValues.forEach(function (v) {
                checkedSet.add(v);
              });
            }
          } else if (checked) {
            if (checkedSet.size >= localMax && isNumber.isNumber_1(max)) return;
            checkedSet.add(checkValue);
          } else {
            checkedSet["delete"](checkValue);
          }

          setInternalValue(Array.from(checkedSet), {
            e: e
          });
        }
      });
    }
  };
  var useOptions = Array.isArray(options) && options.length !== 0;
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-checkbox-group"), className),
    style: style
  }, /* @__PURE__ */React__default["default"].createElement(common_Check.CheckContext.Provider, {
    value: context
  }, useOptions ? options.map(function (v, index) {
    var type = _typeof._typeof(v);

    switch (type) {
      case "number":
        {
          var vs = v;
          return /* @__PURE__ */React__default["default"].createElement(Checkbox, {
            key: vs,
            label: vs,
            value: vs
          }, v);
        }

      case "object":
        {
          var _vs = v;
          return _vs.checkAll ? /* @__PURE__ */React__default["default"].createElement(Checkbox, _objectSpread$1(_objectSpread$1({}, v), {}, {
            key: "checkAll_".concat(index),
            indeterminate: indeterminate
          })) : /* @__PURE__ */React__default["default"].createElement(Checkbox, _objectSpread$1(_objectSpread$1({}, v), {}, {
            key: _vs.value,
            disabled: _vs.disabled || disabled
          }));
        }

      default:
        return null;
    }
  }) : children));
}

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Checkbox = _util_forwardRefWithStatics["default"](function (props, ref) {
  return /* @__PURE__ */React__default["default"].createElement(common_Check["default"], _objectSpread({
    ref: ref,
    type: "checkbox"
  }, props));
}, {
  Group: CheckboxGroup
});
Checkbox.displayName = "Checkbox";

exports.Checkbox = Checkbox;
exports.CheckboxGroup = CheckboxGroup;
