// tdesign-react/cjs/_chunks/dep-e4430646.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

var defineProperty = require('./dep-449daca3.js');
var slicedToArray = require('./dep-fa36e79f.js');
var objectWithoutProperties = require('./dep-e03ca79d.js');
var React = require('react');
var useResizeObserver = require('use-resize-observer');
var classNames = require('classnames');
var _util_useConfig = require('../_util/useConfig.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var _util_useCommonClassName = require('../_util/useCommonClassName.js');
var _util_composeRefs = require('../_util/composeRefs.js');
var avatar_AvatarContext = require('../avatar/AvatarContext.js');
var popup_Popup = require('../popup/Popup.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var useResizeObserver__default = /*#__PURE__*/_interopDefaultLegacy(useResizeObserver);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded$1 = ["className", "cascading", "collapseAvatar", "max", "placement", "popupProps", "size", "children"];

function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys$1(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys$1(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Group = function Group(props) {
  var _classNames;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var preClass = "".concat(classPrefix, "-avatar");

  var className = props.className,
      _props$cascading = props.cascading,
      cascading = _props$cascading === void 0 ? "right-up" : _props$cascading,
      collapseAvatar = props.collapseAvatar,
      max = props.max,
      placement = props.placement,
      popupProps = props.popupProps,
      _props$size = props.size,
      size = _props$size === void 0 ? "medium" : _props$size,
      children = props.children,
      avatarGroupProps = objectWithoutProperties._objectWithoutProperties(props, _excluded$1);

  var childrenList = React__default["default"].Children.toArray(children);
  var allChildrenList;

  if (childrenList.length > 0) {
    allChildrenList = childrenList.map(function (child, index) {
      return /*#__PURE__*/React__default["default"].cloneElement(child, _objectSpread$1({
        key: "avatar-group-item-".concat(index)
      }, child.props));
    });
  }

  var groupClass = classNames__default["default"]("".concat(preClass, "-group"), className, (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(preClass, "--offset-right"), cascading === "right-up"), defineProperty._defineProperty(_classNames, "".concat(preClass, "--offset-left"), cascading === "left-up"), _classNames));
  var childrenCount = childrenList.length;

  if (max && childrenCount > max) {
    var showList = allChildrenList.slice(0, max);
    var hiddenList = allChildrenList.slice(max, childrenCount);
    var popupNum = "+".concat(childrenCount - max);

    var popupMergeProps = _objectSpread$1(_objectSpread$1({}, popupProps), {}, {
      placement: placement
    });

    var popupNodes = popupProps ? /* @__PURE__ */React__default["default"].createElement(popup_Popup["default"], _objectSpread$1({}, popupMergeProps), collapseAvatar ? /* @__PURE__ */React__default["default"].createElement(Avatar, {
      size: size
    }, collapseAvatar) : /* @__PURE__ */React__default["default"].createElement(Avatar, {
      size: size
    }, popupNum)) : /* @__PURE__ */React__default["default"].createElement(popup_Popup["default"], {
      key: "avatar-popup-key",
      placement: placement,
      content: hiddenList,
      trigger: "hover",
      showArrow: true
    }, collapseAvatar ? /* @__PURE__ */React__default["default"].createElement(Avatar, {
      size: size
    }, collapseAvatar) : /* @__PURE__ */React__default["default"].createElement(Avatar, {
      size: size
    }, popupNum));
    showList.push(popupNodes);
    return /* @__PURE__ */React__default["default"].createElement(avatar_AvatarContext.AvatarContextProvider, {
      size: size
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: groupClass
    }, showList));
  }

  return /* @__PURE__ */React__default["default"].createElement(avatar_AvatarContext.AvatarContextProvider, {
    size: size
  }, /* @__PURE__ */React__default["default"].createElement("div", _objectSpread$1({
    className: groupClass
  }, avatarGroupProps), allChildrenList));
};

var _excluded = ["alt", "hideOnLoadFailed", "icon", "image", "shape", "size", "onError", "children", "style", "className"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Avatar = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _classNames;

  var alt = props.alt,
      _props$hideOnLoadFail = props.hideOnLoadFailed,
      hideOnLoadFailed = _props$hideOnLoadFail === void 0 ? false : _props$hideOnLoadFail,
      icon = props.icon,
      image = props.image,
      _props$shape = props.shape,
      shape = _props$shape === void 0 ? "circle" : _props$shape,
      _props$size = props.size,
      avatarSize = _props$size === void 0 ? "default" : _props$size,
      onError = props.onError,
      children = props.children,
      style = props.style,
      className = props.className,
      avatarProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);

  var groupSize = React.useContext(avatar_AvatarContext["default"]);

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useState = React.useState(1),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      scale = _useState2[0],
      setScale = _useState2[1];

  var _useState3 = React.useState(true),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      isImgExist = _useState4[0],
      setIsImgExist = _useState4[1];

  var avatarRef = React.useRef(null);
  var avatarChildrenRef = React.useRef(null);
  var size = avatarSize === "default" ? groupSize : avatarSize;
  var gap = 4;

  var handleScale = function handleScale() {
    if (!avatarChildrenRef.current || !avatarRef.current) {
      return;
    }

    var childrenWidth = avatarChildrenRef.current.offsetWidth;
    var avatarWidth = avatarRef.current.offsetWidth;

    if (childrenWidth !== 0 && avatarWidth !== 0) {
      if (gap * 2 < avatarWidth) {
        setScale(avatarWidth - gap * 2 < childrenWidth ? (avatarWidth - gap * 2) / childrenWidth : 1);
      }
    }
  };

  var _useResizeObserver = useResizeObserver__default["default"]({
    onResize: handleScale
  }),
      observerRef = _useResizeObserver.ref;

  var handleImgLoadError = function handleImgLoadError() {
    onError && onError();
    !hideOnLoadFailed && setIsImgExist(false);
  };

  React.useEffect(function () {
    setIsImgExist(true);
    setScale(1);
  }, [props.image]);
  React.useEffect(function () {
    handleScale();
  }, []);

  var _useCommonClassName = _util_useCommonClassName["default"](),
      SIZE = _useCommonClassName.SIZE;

  var numSizeStyle = size && !SIZE[size] ? {
    width: size,
    height: size,
    fontSize: "".concat(Number.parseInt(size, 10) / 2, "px")
  } : {};
  var imageStyle = size && !SIZE[size] ? {
    width: size,
    height: size
  } : {};
  var preClass = "".concat(classPrefix, "-avatar");
  var avatarClass = classNames__default["default"](preClass, className, (_classNames = {}, defineProperty._defineProperty(_classNames, SIZE[size], !!SIZE[size]), defineProperty._defineProperty(_classNames, "".concat(preClass, "--").concat(shape), !!shape), defineProperty._defineProperty(_classNames, "".concat(preClass, "-icon"), !!icon), _classNames));
  var content;

  if (image && isImgExist) {
    content = /* @__PURE__ */React__default["default"].createElement("img", {
      src: image,
      alt: alt,
      style: imageStyle,
      onError: handleImgLoadError
    });
  } else if (icon) {
    content = icon;
  } else {
    var childrenStyle = {
      transform: "scale(".concat(scale, ")")
    };
    content = /* @__PURE__ */React__default["default"].createElement("span", {
      ref: _util_composeRefs["default"](ref, avatarChildrenRef, observerRef),
      style: childrenStyle
    }, children);
  }

  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    ref: _util_composeRefs["default"](ref, avatarRef),
    className: avatarClass,
    style: _objectSpread(_objectSpread({}, numSizeStyle), style)
  }, avatarProps), content);
}, {
  Group: Group
});
Avatar.displayName = "Avatar";

exports.Avatar = Avatar;
exports.Group = Group;
