// tdesign-react/lib/table/primary/useFilter.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-9332bbea.js';
import { _ as _slicedToArray } from '../../_chunks/dep-93971e86.js';
import { useState, useEffect, useMemo } from 'react';
import FilterButton from './FilterButton.js';
import '../../_chunks/dep-beac373a.js';
import 'classnames';
import '../../_chunks/dep-a347fb20.js';
import '../../_chunks/dep-c6576f8f.js';
import '../../_chunks/dep-156c830d.js';
import '../../_chunks/dep-b33ecffe.js';
import '../../_chunks/dep-f3f6eedc.js';
import '../../_chunks/dep-04305406.js';
import '../../_chunks/dep-6ca758b4.js';
import 'tdesign-icons-react';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../_util/useDefault.js';
import '../../_util/noop.js';
import '../../_util/useConfig.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_util/composeRefs.js';
import '../../_util/usePrevious.js';
import '../../popup/Portal.js';
import 'react-dom';
import '../../popup/hooks/useTriggerProps.js';
import '../../_util/useClickOutside.js';
import '../../popup/hooks/usePopupCssTransition.js';
import '../../radio/index.js';
import '../../_chunks/dep-8610b562.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../common/Check.js';
import '../../_chunks/dep-27d690ce.js';
import '../../_util/helper.js';
import '../../_chunks/dep-5913685b.js';
import '../../_chunks/dep-76c67131.js';
import '../../_chunks/dep-7cf92957.js';
import '../../_chunks/dep-e7eff919.js';
import '../../_chunks/dep-3dccad86.js';
import '../../_util/useCommonClassName.js';
import '../../checkbox/index.js';
import '../../_chunks/dep-18a6adb5.js';
import '../../_chunks/dep-e2a4740c.js';
import '../../input/index.js';
import '../../input/Input.js';
import '../../input/InputGroup.js';
import '../../_util/useDefaultValue.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function useFilter(props) {
  var columns = props.columns,
      filterIcon = props.filterIcon,
      filterValue = props.filterValue,
      defaultFilterValue = props.defaultFilterValue,
      onFilterChange = props.onFilterChange,
      data = props.data;
  var isControlled = typeof filterValue !== "undefined";

  var _useState = useState(defaultFilterValue),
      _useState2 = _slicedToArray(_useState, 2),
      filterVal = _useState2[0],
      setFilterVal = _useState2[1];

  useEffect(function () {
    setFilterVal(isControlled ? filterValue : defaultFilterValue);
  }, [filterValue, defaultFilterValue, isControlled]);
  var transformedFilterData = useMemo(function () {
    if (isControlled) {
      return data;
    }

    if (filterVal) {
      var tmpData = data;
      Object.keys(filterVal).forEach(function (k) {
        if (typeof (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) === "string" || typeof (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) === "number") {
          tmpData = tmpData.filter(function (item) {
            var _item$k;

            return item === null || item === void 0 ? void 0 : (_item$k = item[k]) === null || _item$k === void 0 ? void 0 : _item$k.toString().includes(filterVal === null || filterVal === void 0 ? void 0 : filterVal[k].toString());
          });
        }

        if (Array.isArray(filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) && (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]).length > 0) {
          tmpData = tmpData.filter(function (item) {
            return (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]).includes(item === null || item === void 0 ? void 0 : item[k]);
          });
        }
      });
      return tmpData;
    }

    return data;
  }, [data, filterVal, isControlled]);
  var transformedFilterColumns = FilterButton({
    columns: columns,
    onChange: onChangeFilterButton,
    innerfiltVal: isControlled ? filterValue : defaultFilterValue,
    filterIcon: filterIcon
  });

  function onChangeFilterButton(toast, col) {
    var newFilterVal = _objectSpread(_objectSpread({}, filterVal), toast);

    setFilterVal(newFilterVal);

    if (onFilterChange) {
      onFilterChange(newFilterVal, {
        col: col
      });
    }
  }

  return [transformedFilterColumns, transformedFilterData];
}

export { useFilter as default };
