// tdesign-react/es/table/primary/useFilter.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-9332bbea.js';
import { _ as _slicedToArray } from '../../_chunks/dep-93971e86.js';
import { useState, useEffect, useMemo } from 'react';
import FilterButton from './FilterButton.js';
import '../../_chunks/dep-beac373a.js';
import 'classnames';
import '../../_chunks/dep-25536592.js';
import '../../_chunks/dep-ab1e2688.js';
import '../../_chunks/dep-aa2bff1b.js';
import '../../_chunks/dep-156c830d.js';
import '../../_chunks/dep-a764e3bd.js';
import '../../_chunks/dep-d098d908.js';
import '../../_chunks/dep-3853103d.js';
import '../../_chunks/dep-897ce04d.js';
import '../../_chunks/dep-2eff142e.js';
import '../../_chunks/dep-312619d6.js';
import 'tdesign-icons-react';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../_util/useDefault.js';
import '../../_util/noop.js';
import '../../_util/useConfig.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_util/composeRefs.js';
import '../../_util/usePrevious.js';
import '../../popup/Portal.js';
import 'react-dom';
import '../../popup/hooks/useTriggerProps.js';
import '../../_util/useClickOutside.js';
import '../../popup/hooks/usePopupCssTransition.js';
import './style/css.js';
import '../../popup/type.js';
import '../../radio/index.js';
import '../../_chunks/dep-6cc10247.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../common/Check.js';
import '../../_chunks/dep-27d690ce.js';
import '../../_util/helper.js';
import '../../_chunks/dep-93bd2d05.js';
import '../../_chunks/dep-44c00646.js';
import '../../_chunks/dep-95c9ff5b.js';
import '../../_chunks/dep-304ab091.js';
import '../../_chunks/dep-ea69dcf8.js';
import '../../_util/useCommonClassName.js';
import '../../radio/type.js';
import '../../checkbox/index.js';
import '../../_chunks/dep-d44ac8a4.js';
import '../../_chunks/dep-d2535fd6.js';
import '../../checkbox/type.js';
import '../../input/index.js';
import '../../input/Input.js';
import '../../input/InputGroup.js';
import '../../_util/useDefaultValue.js';
import '../../input/type.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function useFilter(props) {
  var columns = props.columns,
      filterIcon = props.filterIcon,
      filterValue = props.filterValue,
      defaultFilterValue = props.defaultFilterValue,
      onFilterChange = props.onFilterChange,
      data = props.data;
  var isControlled = typeof filterValue !== "undefined";

  var _useState = useState(defaultFilterValue),
      _useState2 = _slicedToArray(_useState, 2),
      filterVal = _useState2[0],
      setFilterVal = _useState2[1];

  useEffect(function () {
    setFilterVal(isControlled ? filterValue : defaultFilterValue);
  }, [filterValue, defaultFilterValue, isControlled]);
  var transformedFilterData = useMemo(function () {
    if (isControlled) {
      return data;
    }

    if (filterVal) {
      var tmpData = data;
      Object.keys(filterVal).forEach(function (k) {
        if (typeof (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) === "string" || typeof (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) === "number") {
          tmpData = tmpData.filter(function (item) {
            var _item$k;

            return item === null || item === void 0 ? void 0 : (_item$k = item[k]) === null || _item$k === void 0 ? void 0 : _item$k.toString().includes(filterVal === null || filterVal === void 0 ? void 0 : filterVal[k].toString());
          });
        }

        if (Array.isArray(filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]) && (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]).length > 0) {
          tmpData = tmpData.filter(function (item) {
            return (filterVal === null || filterVal === void 0 ? void 0 : filterVal[k]).includes(item === null || item === void 0 ? void 0 : item[k]);
          });
        }
      });
      return tmpData;
    }

    return data;
  }, [data, filterVal, isControlled]);
  var transformedFilterColumns = FilterButton({
    columns: columns,
    onChange: onChangeFilterButton,
    innerfiltVal: isControlled ? filterValue : defaultFilterValue,
    filterIcon: filterIcon
  });

  function onChangeFilterButton(toast, col) {
    var newFilterVal = _objectSpread(_objectSpread({}, filterVal), toast);

    setFilterVal(newFilterVal);

    if (onFilterChange) {
      onFilterChange(newFilterVal, {
        col: col
      });
    }
  }

  return [transformedFilterColumns, transformedFilterData];
}

export { useFilter as default };
