// tdesign-react/es/upload/themes/flow-list/file-list.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import React from 'react';
import { CheckCircleFilledIcon, ErrorCircleFilledIcon, TimeFilledIcon } from 'tdesign-icons-react';
import { abridgeName, returnFileSize } from '../../util.js';
import useConfig from '../../../_util/useConfig.js';
import { Loading } from '../../../loading/index.js';
import '../../../_chunks/dep-124f91c5.js';
import '../../../_chunks/dep-beac373a.js';
import '../../../config-provider/ConfigContext.js';
import '../../../locale/zh_CN.js';
import '../../../loading/loading.js';
import '../../../_chunks/dep-9332bbea.js';
import '../../../_chunks/dep-93971e86.js';
import 'classnames';
import '../../../_util/dom.js';
import 'raf';
import '../../../_chunks/dep-3853103d.js';
import '../../../_chunks/dep-ab1e2688.js';
import '../../../_chunks/dep-aa2bff1b.js';
import '../../../_chunks/dep-156c830d.js';
import '../../../_chunks/dep-897ce04d.js';
import '../../../_chunks/dep-2eff142e.js';
import '../../../_chunks/dep-312619d6.js';
import '../../../_util/easing.js';
import '../../../common/Portal.js';
import 'react-dom';
import '../../../loading/gradient.js';
import './style/css.js';
import '../../../loading/type.js';
import '../../../loading/plugin.js';

var FileList = function FileList(props) {
  var listFiles = props.listFiles,
      showInitial = props.showInitial,
      renderDragger = props.renderDragger,
      showUploadProgress = props.showUploadProgress,
      remove = props.remove;

  var _useConfig = useConfig(),
      prefix = _useConfig.classPrefix;

  var UPLOAD_NAME = "".concat(prefix, "-upload");

  var renderStatus = function renderStatus(file) {
    var STATUS_MAP = {
      success: {
        icon: /* @__PURE__ */React.createElement(CheckCircleFilledIcon, null),
        text: "\u4E0A\u4F20\u6210\u529F"
      },
      fail: {
        icon: /* @__PURE__ */React.createElement(ErrorCircleFilledIcon, null),
        text: "\u4E0A\u4F20\u5931\u8D25"
      },
      progress: {
        icon: /* @__PURE__ */React.createElement(Loading, null),
        text: "\u4E0A\u4F20\u4E2D ".concat(Math.min(file.percent, 99), "%")
      },
      waiting: {
        icon: /* @__PURE__ */React.createElement(TimeFilledIcon, null),
        text: "\u5F85\u4E0A\u4F20"
      }
    };
    if (file.status === "progress" && !showUploadProgress) return null;
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(UPLOAD_NAME, "__flow-status")
    }, STATUS_MAP[file.status].icon, /* @__PURE__ */React.createElement("span", null, STATUS_MAP[file.status].text));
  };

  return /* @__PURE__ */React.createElement("table", {
    className: "".concat(UPLOAD_NAME, "__flow-table")
  }, /* @__PURE__ */React.createElement("thead", null, /* @__PURE__ */React.createElement("tr", null, /* @__PURE__ */React.createElement("th", null, "\u6587\u4EF6\u540D"), /* @__PURE__ */React.createElement("th", null, "\u5927\u5C0F"), /* @__PURE__ */React.createElement("th", null, "\u72B6\u6001"), /* @__PURE__ */React.createElement("th", null, "\u64CD\u4F5C"))), /* @__PURE__ */React.createElement("tbody", null, showInitial && /* @__PURE__ */React.createElement("tr", null, /* @__PURE__ */React.createElement("td", {
    colSpan: 4
  }, renderDragger())), listFiles.map(function (file, index) {
    return /* @__PURE__ */React.createElement("tr", {
      key: index
    }, /* @__PURE__ */React.createElement("td", null, abridgeName(file.name, 7, 10)), /* @__PURE__ */React.createElement("td", null, returnFileSize(file.size)), /* @__PURE__ */React.createElement("td", null, renderStatus(file)), /* @__PURE__ */React.createElement("td", null, /* @__PURE__ */React.createElement("span", {
      className: "".concat(UPLOAD_NAME, "__flow-button"),
      onClick: function onClick(e) {
        return remove({
          e: e,
          index: index,
          file: file
        });
      }
    }, "\u5220\u9664")));
  })));
};

export { FileList as default };
