// tdesign-react/lib/date-picker/panel/DateRange.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-93971e86.js';
import React, { useState, useEffect, useCallback, useMemo } from 'react';
import dayjs from 'dayjs';
import useConfig from '../../_util/useConfig.js';
import DatePickerHeader from '../base/Header.js';
import DatePickerTable from '../base/Table.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import { getToday, isSame, addMonth, getYears, getMonths, getWeeks, flagActive, subtractMonth, setDateTime } from '../utils.js';
import '../../_chunks/dep-beac373a.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import 'tdesign-icons-react';
import '../../button/index.js';
import '../../button/Button.js';
import '../../_chunks/dep-9332bbea.js';
import '../../_chunks/dep-27d690ce.js';
import 'classnames';
import '../../_util/noop.js';
import '../../_util/useRipple.js';
import '../../_util/setStyle.js';
import '../../loading/index.js';
import '../../loading/loading.js';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-f3f6eedc.js';
import '../../_chunks/dep-c6576f8f.js';
import '../../_chunks/dep-156c830d.js';
import '../../_chunks/dep-04305406.js';
import '../../_chunks/dep-6ca758b4.js';
import '../../_util/easing.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../loading/plugin.js';
import '../base/Cell.js';
import '../../_chunks/dep-ab9c749e.js';
import '../../_chunks/dep-3dccad86.js';
import '../../_chunks/dep-5950eb93.js';
import '../../_chunks/dep-0268bf88.js';
import '../../_chunks/dep-aeb2fb38.js';
import '../../_chunks/dep-a347fb20.js';
import '../../_chunks/dep-b33ecffe.js';
import '../../_chunks/dep-17892822.js';
import '../../_chunks/dep-99f6ab64.js';
import '../../_chunks/dep-61d30284.js';
import '../../_chunks/dep-fc84e5e6.js';
import '../../_chunks/dep-7cf92957.js';

var TODAY = getToday();
var LEFT = "left";
var RIGHT = "right";

var DateRangePanel = function DateRangePanel(props) {
  var _useLocaleReceiver = useLocaleReceiver("datePicker"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var monthAriaLabel = t(local.monthAriaLabel);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var value = props.value,
      mode = props.mode,
      minDate = props.minDate,
      maxDate = props.maxDate,
      firstDayOfWeek = props.firstDayOfWeek,
      disableDate = props.disableDate,
      onChange = props.onChange;
  var data = getLeftAndRightDataFromValue(value);

  var _useState = useState(data.leftYear),
      _useState2 = _slicedToArray(_useState, 2),
      leftYear = _useState2[0],
      setLeftYear = _useState2[1];

  var _useState3 = useState(data.leftMonth),
      _useState4 = _slicedToArray(_useState3, 2),
      leftMonth = _useState4[0],
      setLeftMonth = _useState4[1];

  var _useState5 = useState(data.rightYear),
      _useState6 = _slicedToArray(_useState5, 2),
      rightYear = _useState6[0],
      setRightYear = _useState6[1];

  var _useState7 = useState(data.rightMonth),
      _useState8 = _slicedToArray(_useState7, 2),
      rightMonth = _useState8[0],
      setRightMonth = _useState8[1];

  var _useState9 = useState(mode),
      _useState10 = _slicedToArray(_useState9, 2),
      leftType = _useState10[0],
      setLeftType = _useState10[1];

  var _useState11 = useState(mode),
      _useState12 = _slicedToArray(_useState11, 2),
      rightType = _useState12[0],
      setRightType = _useState12[1];

  var _useState13 = useState(data.startValue),
      _useState14 = _slicedToArray(_useState13, 2),
      startValue = _useState14[0],
      setStartValue = _useState14[1];

  var _useState15 = useState(data.endValue),
      _useState16 = _slicedToArray(_useState15, 2),
      endValue = _useState16[0],
      setEndValue = _useState16[1];

  var _useState17 = useState(true),
      _useState18 = _slicedToArray(_useState17, 2),
      isFirstClick = _useState18[0],
      setIsFirstClick = _useState18[1];

  var _useState19 = useState(TODAY),
      _useState20 = _slicedToArray(_useState19, 2),
      firstClickValue = _useState20[0],
      setFirstClickValue = _useState20[1];

  useEffect(function () {
    var panelData = getLeftAndRightDataFromValue(value);
    var leftYear2 = panelData.leftYear,
        leftMonth2 = panelData.leftMonth,
        rightMonth2 = panelData.rightMonth,
        rightYear2 = panelData.rightYear,
        startValue2 = panelData.startValue,
        endValue2 = panelData.endValue;
    setStartValue(startValue2);
    setEndValue(endValue2);
    setLeftYear(leftYear2);
    setLeftMonth(leftMonth2);
    setRightYear(rightYear2);
    setRightMonth(rightMonth2);
  }, [value]);
  useEffect(function () {
    setLeftType(mode);
    setRightType(mode);
  }, [mode]);

  function getLeftAndRightDataFromValue(value2) {
    var _value = _slicedToArray(value2, 2),
        _value$ = _value[0],
        startValue2 = _value$ === void 0 ? TODAY : _value$,
        _value$2 = _value[1],
        endValue2 = _value$2 === void 0 ? TODAY : _value$2;

    var leftYear2 = startValue2.getFullYear();
    var leftMonth2 = startValue2.getMonth();
    var rightMonth2 = endValue2.getMonth();
    var rightYear2 = endValue2.getFullYear();

    if (mode === "date" && isSame(startValue2, endValue2, "month")) {
      var next = addMonth(endValue2, 1);
      rightMonth2 = addMonth(endValue2, 1).getMonth();
      rightYear2 = next.getFullYear();
    }

    if (mode === "month" && isSame(startValue2, endValue2, "year")) {
      rightYear2 = leftYear2 + 1;
    }

    if (mode === "year" && isSame(startValue2, endValue2, "year")) {
      rightYear2 = leftYear2 + 10;
    }

    return {
      leftYear: leftYear2,
      leftMonth: leftMonth2,
      rightMonth: rightMonth2,
      rightYear: rightYear2,
      startValue: startValue2,
      endValue: endValue2
    };
  }

  function clickHeader(flag, direction) {
    var year = direction === LEFT ? leftYear : rightYear;
    var month = direction === LEFT ? leftMonth : rightMonth;
    var type = direction === LEFT ? leftType : rightType;
    var monthCount;
    var next;

    switch (type) {
      case "date":
        monthCount = 1;
        break;

      case "month":
        monthCount = 12;
        break;

      case "year":
        monthCount = 120;
    }

    var current = new Date(year, month);

    if (flag === 1) {
      next = addMonth(current, monthCount);
    } else if (flag === -1) {
      next = subtractMonth(current, monthCount);
    } else {
      next = new Date();
    }

    direction === LEFT ? setLeftYear(next.getFullYear()) : setRightYear(next.getFullYear());
    direction === LEFT ? setLeftMonth(next.getMonth()) : setRightMonth(next.getMonth());
  }

  function getClickHandler(direction) {
    var type = direction === LEFT ? leftType : rightType;
    if (type === "date") return function (date) {
      return clickDate(date);
    };
    if (type === "month") return function (date) {
      return clickMonth(date, direction);
    };
    if (type === "year") return function (date) {
      return clickYear(date, direction);
    };
  }

  function clickDate(date) {
    if (isFirstClick) {
      setStartValue(date);
      setEndValue(date);
      setIsFirstClick(false);
      setFirstClickValue(date);
    } else {
      if (dayjs(firstClickValue).isBefore(dayjs(date), "day")) {
        setEndValue(date);
      } else {
        setEndValue(firstClickValue);
        setStartValue(date);
      }

      onChange === null || onChange === void 0 ? void 0 : onChange([setDateTime(startValue, 0, 0, 0), setDateTime(endValue, 23, 59, 59)]);
      setIsFirstClick(true);
    }
  }

  function clickYear(date, direction) {
    if (mode === "year") {
      if (isFirstClick) {
        setStartValue(date);
        setIsFirstClick(false);
        setFirstClickValue(date);
      } else {
        onChange === null || onChange === void 0 ? void 0 : onChange([startValue, endValue]);
        setIsFirstClick(true);
      }
    } else {
      direction === LEFT ? setLeftType("month") : setRightType("month");
      direction === LEFT ? setLeftYear(date.getFullYear()) : setRightYear(date.getFullYear());
    }
  }

  function clickMonth(date, direction) {
    if (mode === "month") {
      if (isFirstClick) {
        setStartValue(date);
        setIsFirstClick(false);
        setFirstClickValue(date);
      } else {
        if (endValue < startValue) {
          setEndValue(startValue);
        }

        onChange === null || onChange === void 0 ? void 0 : onChange([startValue, endValue]);
        setIsFirstClick(true);
      }
    } else {
      direction === LEFT ? setLeftType("date") : setRightType("date");
      direction === LEFT ? setLeftYear(date.getFullYear()) : setRightYear(date.getFullYear());
      direction === LEFT ? setLeftMonth(date.getMonth()) : setRightMonth(date.getMonth());
    }
  }

  function onMouseEnter(date) {
    if (isFirstClick) return;

    if (firstClickValue.getTime() > date.getTime()) {
      setStartValue(date);
      setEndValue(firstClickValue);
    } else {
      setStartValue(firstClickValue);
      setEndValue(date);
    }
  }

  var getData = useCallback(function (_ref) {
    var year = _ref.year,
        month = _ref.month,
        type = _ref.type,
        start = _ref.start,
        end = _ref.end;
    var data2;
    var options = {
      disableDate: disableDate,
      minDate: minDate,
      maxDate: maxDate,
      firstDayOfWeek: firstDayOfWeek,
      monthText: monthAriaLabel
    };

    switch (type) {
      case "date":
        data2 = getWeeks({
          year: year,
          month: month
        }, options);
        break;

      case "month":
        data2 = getMonths(year, options);
        break;

      case "year":
        data2 = getYears(year, options);
        break;
    }

    return flagActive(data2, {
      start: start,
      end: end,
      type: type
    });
  }, [disableDate, minDate, maxDate, firstDayOfWeek, monthAriaLabel]);
  var leftData = useMemo(function () {
    return getData({
      year: leftYear,
      month: leftMonth,
      type: leftType,
      start: startValue,
      end: endValue
    });
  }, [leftYear, leftMonth, leftType, startValue, endValue, getData]);
  var rightData = useMemo(function () {
    return getData({
      year: rightYear,
      month: rightMonth,
      type: rightType,
      start: startValue,
      end: endValue
    });
  }, [rightYear, rightMonth, rightType, startValue, endValue, getData]);
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-date-picker__panels")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-date-picker__panel")
  }, /* @__PURE__ */React.createElement(DatePickerHeader, {
    year: leftYear,
    month: leftMonth,
    type: leftType,
    onBtnClick: function onBtnClick(flag) {
      return clickHeader(flag, LEFT);
    },
    onTypeChange: setLeftType
  }), /* @__PURE__ */React.createElement(DatePickerTable, {
    type: leftType,
    firstDayOfWeek: firstDayOfWeek,
    data: leftData,
    onCellClick: getClickHandler(LEFT),
    onCellMouseEnter: onMouseEnter
  })), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-date-picker__panel")
  }, /* @__PURE__ */React.createElement(DatePickerHeader, {
    year: rightYear,
    month: rightMonth,
    type: rightType,
    onBtnClick: function onBtnClick(flag) {
      return clickHeader(flag, RIGHT);
    },
    onTypeChange: setRightType
  }), /* @__PURE__ */React.createElement(DatePickerTable, {
    type: rightType,
    firstDayOfWeek: firstDayOfWeek,
    data: rightData,
    onCellClick: getClickHandler(RIGHT),
    onCellMouseEnter: onMouseEnter
  })));
};

DateRangePanel.displayName = "DateRangePanel";
DateRangePanel.defaultProps = {
  value: [TODAY, TODAY],
  mode: "date"
};

export { DateRangePanel as default };
