// tdesign-react/cjs/time-picker/consts.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-fa36e79f.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
require('../_chunks/dep-045b87c1.js');
require('../_chunks/dep-449daca3.js');
require('react');
require('../config-provider/ConfigContext.js');
require('../locale/zh_CN.js');

var useTimePickerTextConfig = function useTimePickerTextConfig() {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("timePicker"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  return {
    nowtime: t(local.nowtime),
    confirm: t(local.confirm),
    am: t(local.anteMeridiem),
    pm: t(local.postMeridiem),
    placeholder: t(local.placeholder)
  };
};
var AM = "am";
var PM = "pm";
var MERIDIEM_LIST = [AM, PM];
var DEFAULT_STEPS = [1, 1, 1];
var DEFAULT_FORMAT = "HH:mm:ss";

exports.AM = AM;
exports.DEFAULT_FORMAT = DEFAULT_FORMAT;
exports.DEFAULT_STEPS = DEFAULT_STEPS;
exports.MERIDIEM_LIST = MERIDIEM_LIST;
exports.PM = PM;
exports.useTimePickerTextConfig = useTimePickerTextConfig;
