// tdesign-react/esm/time-picker/panel/TimePickerRangePanel.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-9332bbea.js';
import { _ as _slicedToArray } from '../../_chunks/dep-93971e86.js';
import React, { useMemo } from 'react';
import classNames from 'classnames';
import dayjs from 'dayjs';
import SinglePanel from './SinglePanel.js';
import useConfig from '../../_util/useConfig.js';
import { Button } from '../../button/index.js';
import { DEFAULT_STEPS, DEFAULT_FORMAT, useTimePickerTextConfig } from '../consts.js';
import '../../_chunks/dep-beac373a.js';
import '../../_chunks/dep-9e88086c.js';
import '../../_chunks/dep-ab1e2688.js';
import '../../_chunks/dep-a764e3bd.js';
import '../../_chunks/dep-156c830d.js';
import '../../_chunks/dep-d098d908.js';
import '../../_chunks/dep-aa2bff1b.js';
import '../../_chunks/dep-cac5f189.js';
import '../../_chunks/dep-95c9ff5b.js';
import '../../_chunks/dep-312619d6.js';
import '../../_chunks/dep-8dc257e5.js';
import '../../_chunks/dep-44c00646.js';
import '../../_chunks/dep-897ce04d.js';
import '../../_chunks/dep-2eff142e.js';
import '../../_chunks/dep-304ab091.js';
import '../../_chunks/dep-ea69dcf8.js';
import '../../_chunks/dep-8bccb9b0.js';
import '../../_chunks/dep-028895a6.js';
import '../../_chunks/dep-6bae5139.js';
import '../../_chunks/dep-0de37580.js';
import '../../_chunks/dep-b8263a1c.js';
import '../../_chunks/dep-25536592.js';
import '../../_chunks/dep-8309e373.js';
import '../../_chunks/dep-14692160.js';
import '../../_chunks/dep-f88db21e.js';
import '../../_util/noop.js';
import '../interfaces.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../button/Button.js';
import '../../_chunks/dep-27d690ce.js';
import '../../_util/useRipple.js';
import '../../_util/setStyle.js';
import '../../loading/index.js';
import '../../loading/loading.js';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-3853103d.js';
import '../../_util/easing.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../loading/style/index.js';
import '../../loading/type.js';
import '../../loading/plugin.js';
import '../../button/style/index.js';
import '../../button/type.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var TimePickerPanel = function TimePickerPanel(props) {
  var _props$value = props.value,
      value = _props$value === void 0 ? [] : _props$value,
      onChange = props.onChange,
      isFooterDisplay = props.isFooterDisplay,
      handleConfirmClick = props.handleConfirmClick,
      _props$steps = props.steps,
      steps = _props$steps === void 0 ? DEFAULT_STEPS : _props$steps,
      _props$format = props.format,
      format = _props$format === void 0 ? DEFAULT_FORMAT : _props$format;
  var TEXT_CONFIG = useTimePickerTextConfig();

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _ref = value || [],
      _ref2 = _slicedToArray(_ref, 2),
      startTime = _ref2[0],
      endTime = _ref2[1];

  var panelClassName = "".concat(classPrefix, "-time-picker__panel");

  var handlePanelValueChange = function handlePanelValueChange(value2, index) {
    if (index === 0) {
      onChange([value2, endTime || value2]);
    } else {
      var calStartTime = startTime || dayjs().hour(0).minute(0).second(0).format(format);
      onChange([calStartTime, value2]);
    }
  };

  var defaultValue = useMemo(function () {
    if (value && value.length === 0) {
      return [dayjs().format(format), dayjs().format(format)];
    }

    return value;
  }, [value, format]);
  return /* @__PURE__ */React.createElement("div", {
    className: classNames(panelClassName, "".concat(panelClassName, "-section"))
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelClassName, "-section-body")
  }, /* @__PURE__ */React.createElement(SinglePanel, _objectSpread(_objectSpread({}, props), {}, {
    steps: steps,
    format: format,
    value: startTime,
    onChange: function onChange(v) {
      return handlePanelValueChange(v, 0);
    }
  })), /* @__PURE__ */React.createElement(SinglePanel, _objectSpread(_objectSpread({}, props), {}, {
    steps: steps,
    format: format,
    value: endTime,
    onChange: function onChange(v) {
      return handlePanelValueChange(v, 1);
    }
  }))), isFooterDisplay ? /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelClassName, "-section-footer")
  }, /* @__PURE__ */React.createElement(Button, {
    theme: "primary",
    variant: "base",
    onClick: function onClick() {
      handleConfirmClick(defaultValue);
    }
  }, TEXT_CONFIG.confirm)) : null);
};

export { TimePickerPanel as default };
