// tdesign-react/lib/table/base/TableAsyncLoadingBody.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

import React from 'react';
import { i as isFunction_1 } from '../../_chunks/dep-a347fb20.js';
import { Loading } from '../../loading/index.js';
import useConfig from '../../_util/useConfig.js';
import '../../_chunks/dep-c6576f8f.js';
import '../../_chunks/dep-156c830d.js';
import '../../_chunks/dep-b33ecffe.js';
import '../../loading/loading.js';
import '../../_chunks/dep-9332bbea.js';
import '../../_chunks/dep-93971e86.js';
import '../../_chunks/dep-beac373a.js';
import 'classnames';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-f3f6eedc.js';
import '../../_chunks/dep-04305406.js';
import '../../_chunks/dep-6ca758b4.js';
import '../../_util/easing.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../loading/plugin.js';

function TableEmptyBody(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var asyncLoading = props.asyncLoading,
      onAsyncLoadingClick = props.onAsyncLoadingClick;
  if (!asyncLoading) return null;

  if (asyncLoading === "loading") {
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-table__async-loading ").concat(classPrefix, "-is-loading"),
      onClick: function onClick() {
        onAsyncLoadingClick({
          status: "loading"
        });
      }
    }, /* @__PURE__ */React.createElement(Loading, {
      loading: true,
      size: "small"
    }), /* @__PURE__ */React.createElement("span", null, "\u6B63\u5728\u52A0\u8F7D\u4E2D\uFF0C\u8BF7\u7A0D\u540E"));
  }

  if (asyncLoading === "load-more") {
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-table__async-loading ").concat(classPrefix, "-is-load-more"),
      onClick: function onClick() {
        onAsyncLoadingClick({
          status: "load-more"
        });
      }
    }, "\u70B9\u51FB\u52A0\u8F7D\u66F4\u591A");
  }

  var result = null;
  if (asyncLoading) result = asyncLoading;
  if (isFunction_1(asyncLoading)) result = asyncLoading();
  return /* @__PURE__ */React.createElement(React.Fragment, null, result);
}

export { TableEmptyBody as default };
