// tdesign-react/cjs/tree-select/useTreeSelectConfig.js
/**
 * tdesign v0.22.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-449daca3.js');
var React = require('react');
var classNames = require('classnames');
var _util_useConfig = require('../_util/useConfig.js');
var _util_useCommonClassName = require('../_util/useCommonClassName.js');
require('../config-provider/ConfigContext.js');
require('../locale/zh_CN.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var defaultPopupProps = {
  trigger: "click",
  placement: "bottom-left",
  overlayClassName: "",
  overlayStyle: function overlayStyle(trigger) {
    return {
      width: "".concat(trigger.offsetWidth, "px")
    };
  }
};
function useTreeSelectConfig(props) {
  var _useConfig = _util_useConfig["default"](),
      prefix = _useConfig.classPrefix;

  var CLASSNAMES = _util_useCommonClassName["default"]();
  var disabled = props.disabled,
      prefixIcon = props.prefixIcon,
      size = props.size,
      visible = props.visible,
      popupProps = props.popupProps;
  return React.useMemo(function () {
    var _classNames;

    var selectClassName = classNames__default["default"]("".concat(prefix, "-select"), (_classNames = {}, defineProperty._defineProperty(_classNames, CLASSNAMES.STATUS.disabled, disabled), defineProperty._defineProperty(_classNames, CLASSNAMES.STATUS.active, visible), defineProperty._defineProperty(_classNames, CLASSNAMES.SIZE[size], size), defineProperty._defineProperty(_classNames, "".concat(prefix, "-has-prefix"), prefixIcon), _classNames));
    var popupObject = Object.assign(defaultPopupProps, popupProps);
    var popupClassName = classNames__default["default"](popupObject.overlayClassName, "".concat(prefix, "-select__dropdown"), "narrow-scrollbar");
    return {
      selectClassName: selectClassName,
      popupClassName: popupClassName,
      popupObject: popupObject
    };
  }, [prefix, size, prefixIcon, disabled, visible, popupProps]);
}

exports["default"] = useTreeSelectConfig;
